---
title: "Zhu2019_DataAnalysis"
output:
  html_document: default
    df_print: paged
    toc: true
    toc_float: true
---

```{r, include=FALSE}
knitr::opts_chunk$set(fig.width = 15, fig.height = 5, warning=FALSE, message=FALSE,
                      root.dir = "~/Projects/MetaIBS/scripts/analysis-individual/Zhu-2019/")
```




***********
# 1. IMPORT
***********

### 1.1. Librairies
```{r library-import, echo=FALSE}
library(phyloseq) # for phyloseq object
library(ggplot2)
library(ggsignif)
library(cowplot)
library(plyr)
library(dplyr)
library("plotly") # plot 3D
library("microbiome") # for centered log-ratio
library("coda") # Aitchison distance
library("coda.base") # Aitchison distance
library("vegan") # NMDS
library(pheatmap) # for heatmap
```


### 1.2. Data
```{r import-data}
# Set path
path.root <- "~/Projects/MetaIBS"
path.zhu  <- file.path(path.root, "data/preprocessing/analysis-individual/Zhu-2019/03_EDA-Zhu")

# Import phyloseq object
physeq.zhu <- readRDS(file.path(path.root, "data/phyloseq-objects/physeq_zhu.rds"))

# Sanity check
physeq.zhu
```

Phylogenetic tree was computed with the package `phangorn`, and the script was run on a cluster. Let's check we have correctly generated a phylogenetic tree.

```{r plot-tree, fig.height = 6, fig.width = 10}
# Look at the tree
plot_tree(physeq.zhu, color = "Phylum", ladderize="left")
```


```{r import-for-html, echo = FALSE}
#__________________________________________________________________________
#____________ THIS IS ONLY FOR (quicker) PDF/HTML OUTPUT __________________
#__________________________________________________________________________

# Import phyloseq objects for computing distances
physeq.rel    <- readRDS(file.path(path.zhu, "physeq_relative.rds"))
physeq.clr    <- readRDS(file.path(path.zhu, "physeq_clr.rds"))
physeq.CSN    <- readRDS(file.path(path.zhu, "physeq_CSN.rds"))
physeq.NZcomp <- readRDS(file.path(path.zhu, "physeq_NZcomp.rds"))
```




*****************
# 2. DEMOGRAPHICS
*****************

This dataset has covariates on gender & age. Let's verify the distribution is similar between healthy & IBS groups.

```{r demographics-stats}
# Number of individuals in each group
metadata <- data.frame(sample_data(physeq.zhu))
metadata %>%
  count(host_disease)

# Age
metadata %>%
  group_by(host_disease) %>%
  summarize(mean_age=mean(host_age), sd_age=sd(host_age))
wilcox.test(metadata[metadata$host_disease == "IBS", ]$host_age,
            metadata[metadata$host_disease == "Healthy", ]$host_age) # p=0.001

# Gender
metadata %>%
  count(host_disease, host_sex)
chisq.test(data.frame("Female" = c(6,5),
                      "Male" = c(8,10))) # p=0.88
```




***************
# 3. ABUNDANCES
***************

### 3.1. Absolute abundances
```{r plot-absolute-abundance, fig.width=10, fig.height = 5}
# Plot Phylum
plot_bar(physeq.zhu, fill = "Phylum") + facet_wrap("host_disease", scales="free_x") +
  theme_cowplot()+
  theme(axis.text.x = element_text(size=6, angle=90))+
  labs(x = "Samples", y = "Absolute abundance", title = "Zhu dataset (2019)")
# ggsave(file.path(path.zhu, "absAbundance_phylum.jpg"), width=10, height=5)

# Plot Class
plot_bar(physeq.zhu, fill = "Class")+ facet_wrap("host_disease", scales="free_x") +
  theme_cowplot()+
  theme(axis.text.x = element_blank())+
  labs(x = "Samples", y = "Absolute abundance", title = "Zhu dataset (2019)")
```

**Sequencing depth** characteristics of the Zhu dataset:  
- minimum of `r min(sample_sums(physeq.zhu))` total count per sample  
- median: `r median(sample_sums(physeq.zhu))` total count per sample  
- maximum of `r max(sample_sums(physeq.zhu))` total count per sample  


### 3.2. Relative abundances
```{r plot-relative-abundances, fig.width=10, fig.height = 5, message=FALSE}
# Agglomerate to phylum & class levels
phylum.table <- physeq.zhu %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

class.table <- physeq.zhu %>%
  tax_glom(taxrank = "Class") %>%
  transform_sample_counts(function(x) {x/sum(x)} ) %>%
  psmelt()


# Plot relative abundances
ggplot(phylum.table, aes(x = reorder(Sample, Sample, function(x) mean(phylum.table[Sample == x & Phylum == 'Firmicutes', 'Abundance'])),
                         y = Abundance, fill = Phylum))+
  facet_wrap(~ host_disease, scales = "free_x") + # scales = "free" removes empty lines
  geom_bar(stat = "identity", width=0.95) +
  theme_cowplot()+
  theme(axis.text.x = element_text(size=6, angle=90))+
  labs(x = "Samples", y = "Relative abundance", title = "Zhu dataset (2019)")
# ggsave(file.path(path.zhu, "relAbundance_phylum.jpg"), width=10, height=5)

ggplot(class.table, aes(x = reorder(Sample, Sample, function(x) mean(class.table[Sample == x & Phylum == 'Firmicutes', 'Abundance'])),
                        y = Abundance, fill = Class))+
  facet_wrap(~ host_disease, scales = "free_x") +
  geom_bar(stat = "identity", width=0.95) +
  theme_cowplot()+
  theme(axis.text.x = element_text(size=6, angle=90))+
  labs(x = "Samples", y = "Relative abundance", title = "Zhu dataset (2019)")
# ggsave(file.path(path.zhu, "relAbundance_class.jpg"), width=12, height=5)
```


### 3.3. Firmicutes/Bacteroidota ratio
```{r plot-log-ratio, fig.height = 5, fig.width=3}
# Extract abundance of only Bacteroidota and Firmicutes
bacter <- phylum.table %>%
  filter(Phylum == "Bacteroidota") %>%
  select(c('Sample', 'Abundance', 'host_disease', 'host_sex', 'Phylum')) %>%
  arrange(Sample)

firmi <- phylum.table %>%
  filter(Phylum == "Firmicutes") %>%
  select(c('Sample', 'Abundance', 'host_disease', 'host_sex', 'Phylum')) %>%
  arrange(Sample)

# Calculate log2 ratio Firmicutes/Bacteroidota
ratio.FB <- data.frame('Sample' = bacter$Sample,
                       'host_disease' = bacter$host_disease,
                       'Bacteroidota' = bacter$Abundance,
                       'Firmicutes' = firmi$Abundance,
                       'host_sex' = bacter$host_sex)
ratio.FB$logRatioFB <- log2(ratio.FB$Firmicutes / ratio.FB$Bacteroidota)

# Plot log2 ratio Firmicutes/Bacteroidota
ggplot(ratio.FB, aes(x = host_disease, y = logRatioFB))+
  geom_violin(aes(fill=host_disease))+
  scale_fill_manual(values=scales::alpha(c("blue", "red"), .3))+
  geom_jitter(width=0.1)+
  geom_signif(comparisons = list(c("Healthy", "IBS")), map_signif_level = TRUE, test="wilcox.test") +
  labs(x = "",  y = 'Log2(Firmicutes/Bacteroidota)', title = "Firmicutes:Bacteroidota ratio")+
  theme_cowplot()+
  theme(legend.position="none")
# ggsave(file.path(path.zhu, "ratioFB.jpg"), width=4, height=6)

# Statistical test
wilcox.test(ratio.FB[ratio.FB$host_disease == "IBS","logRatioFB"],
            ratio.FB[ratio.FB$host_disease == "Healthy","logRatioFB"]) # p = 0.009
```




*******************
# 4. NORMALIZE DATA
*******************

```{r normalize-data, eval = FALSE, echo = TRUE}
# Sanity check no sample with less than 500 total count
table(sample_sums(physeq.zhu)<500) # all FALSE


#____________________________________________________________________
# PHYLOSEQ OBJECT WITH NON-ZERO COMPOSITIONS
physeq.NZcomp <- physeq.zhu
otu_table(physeq.NZcomp)[otu_table(physeq.NZcomp) == 0] <- 0.5 # pseudocounts

# Sanity check that 0 values have been replaced
# otu_table(physeq.zhu)[1:10,1:10]
# otu_table(physeq.NZcomp)[1:10,1:10]

# transform into compositions
physeq.NZcomp <- transform_sample_counts(physeq.NZcomp, function(x) x / sum(x) )
table(rowSums(otu_table(physeq.NZcomp))) # check if there is any row not summing to 1

# Save object
saveRDS(physeq.NZcomp, file.path(path.zhu, "physeq_NZcomp.rds"))


#____________________________________________________________________
# PHYLOSEQ OBJECT WITH RELATIVE COUNT (BETWEEN 0 AND 1)
physeq.rel <- physeq.zhu
physeq.rel <- transform_sample_counts(physeq.rel, function(x) x / sum(x) ) # divide each count by the total number of counts (per sample)

# check the counts are all relative
# otu_table(physeq.zhu)[1:5, 1:5]
# otu_table(physeq.rel)[1:5, 1:5]

# sanity check
table(rowSums(otu_table(physeq.rel))) # check if there is any row not summing to 1

# save the physeq.rel object
saveRDS(physeq.rel, file.path(path.zhu, "physeq_relative.rds"))


#____________________________________________________________________
# PHYLOSEQ OBJECT WITH COMMON-SCALE NORMALIZATION
physeq.CSN <- physeq.zhu
physeq.CSN <- transform_sample_counts(physeq.CSN, function(x) (x*min(sample_sums(physeq.CSN))) / sum(x) )

# sanity check
table(rowSums(otu_table(physeq.CSN))) # check that all rows are summing to the same total

# save the physeq.CSN object
saveRDS(physeq.CSN, file.path(path.zhu, "physeq_CSN.rds"))


#____________________________________________________________________
# PHYLOSEQ OBJECT WITH CENTERED LOG RATIO COUNT
physeq.clr <- physeq.zhu
physeq.clr <- microbiome::transform(physeq.clr, "clr") # the function adds pseudocounts itself

# Compare the otu tables in the original phyloseq object and the new one after CLR transformation
# otu_table(physeq.zhu)[1:5, 1:5] # should contain absolute counts
# otu_table(physeq.clr)[1:5, 1:5] # should all be relative

# save the physeq.rel object
saveRDS(physeq.clr, file.path(path.zhu, "physeq_clr.rds"))
```




**********************
# 5. COMPUTE DISTANCES
**********************

### 5.1. UniFrac, Aitchison, Bray-Curtis and Canberra
First, let's look at these four distances of interest.

```{r dist-functions, fig.width = 15, fig.height = 4}
#____________________________________________________________________________________
# Measure distances
getDistances <- function(){
  set.seed(123) # for unifrac, need to set a seed
  glom.UniF <- UniFrac(physeq.rel, weighted=TRUE, normalized=TRUE) # weighted unifrac
  glom.ait  <- phyloseq::distance(physeq.clr, method = 'euclidean') # aitchison
  glom.bray <- phyloseq::distance(physeq.CSN, method = "bray") # bray-curtis
  glom.can  <- phyloseq::distance(physeq.NZcomp, method = "canberra") # canberra
  dist.list <- list("UniF" = glom.UniF, "Ait" = glom.ait, "Canb" = glom.can, "Bray" = glom.bray)
  
  return(dist.list)
}


#____________________________________________________________________________________
# Plot in 2D the distances
plotDistances2D <- function(dlist, ordination="MDS"){
  plist <- NULL
  plist <- vector("list", 4)
  names(plist) <- c("Weighted Unifrac", "Aitchison", "Bray-Curtis", "Canberra")
  
  print("Unifrac")
  # Weighted UniFrac
  set.seed(123)
  iMDS.UniF <- ordinate(physeq.rel, ordination, distance=dlist$UniF)
  plist[[1]] <- plot_ordination(physeq.rel, iMDS.UniF, color="host_disease")
  
  print("Aitchison")
  # Aitchison
  set.seed(123)
  iMDS.Ait <- ordinate(physeq.clr, ordination, distance=dlist$Ait)
  plist[[2]] <- plot_ordination(physeq.clr, iMDS.Ait, color="host_disease")
  
  print("Bray")
  # Bray-Curtis
  set.seed(123)
  iMDS.Bray <- ordinate(physeq.CSN, ordination, distance=dlist$Bray)
  plist[[3]] <- plot_ordination(physeq.CSN, iMDS.Bray, color="host_disease")
  
  print("Canberra")
  # Canberra
  set.seed(123)
  iMDS.Can <- ordinate(physeq.NZcomp, ordination, distance=dlist$Can)
  plist[[4]] <- plot_ordination(physeq.NZcomp, iMDS.Can, color="host_disease")
  
  # Creating a dataframe to plot everything
  plot.df = ldply(plist, function(x) x$data)
  names(plot.df)[1] <- "distance"
  
  return(plot.df)
}
```

Now let's plot!

```{r dist-run}
# Get the distances & the plot data
dist.zhu <- getDistances()
plot.df <- plotDistances2D(dist.zhu)

# Plot
ggplot(plot.df, aes(Axis.1, Axis.2, color=host_disease))+
  geom_point(size=4, alpha=0.5)  + scale_color_manual(values = c('blue', 'red'))+
  facet_wrap(distance~., scales='free', nrow=1)+
  theme_bw()+
  theme(strip.text.x = element_text(size=20))+
  labs(color="Disease")
# ggsave(file.path(path.zhu, "distances4_MDS.jpg"), height = 4, width = 15)
```


### 5.2. Compute distances on families
As there is an anomaly in the ASVs (the ASVs in the healthy samples are different from the ones in the IBS samples), this creates a bias in the distances between samples. As there are many unknown genera (185/1051), we will agglomerate at the family level and compute distances again.

```{r dist-run-family}
# ++ AGGLOMERATE TO FAMILY LEVEL ++
physeq.zhu.fam <- tax_glom(physeq.zhu, taxrank="Family")


# ++ NORMALIZATION ++
physeq.NZcomp.fam <- physeq.zhu.fam
otu_table(physeq.NZcomp.fam)[otu_table(physeq.NZcomp.fam) == 0] <- 0.5
physeq.NZcomp.fam <- transform_sample_counts(physeq.NZcomp.fam, function(x) x / sum(x) )
# table(rowSums(otu_table(physeq.NZcomp.fam))) # sanity check

physeq.rel.fam <- transform_sample_counts(physeq.zhu.fam, function(x) x / sum(x) )
# table(rowSums(otu_table(physeq.rel.fam))) # sanity check

physeq.CSN.fam <- transform_sample_counts(physeq.zhu.fam, function(x) (x*min(sample_sums(physeq.zhu.fam))) / sum(x) )
# table(rowSums(otu_table(physeq.CSN.fam))) # sanity check

physeq.clr.fam <- microbiome::transform(physeq.zhu.fam, "clr")


# ++ COMPUTE DISTANCES ++
set.seed(123)
dist.zhu.fam <- list("UniF" = UniFrac(physeq.rel.fam, weighted=TRUE, normalized=TRUE),
                     "Ait"  = phyloseq::distance(physeq.clr.fam, method = 'euclidean'),
                     "Canb" = phyloseq::distance(physeq.CSN.fam, method = "bray"),
                     "Bray" = phyloseq::distance(physeq.NZcomp.fam, method = "canberra"))


# ++ PLOT MDS ++
plist <- vector("list", 4)
names(plist) <- c("Weighted Unifrac", "Aitchison", "Bray-Curtis", "Canberra")
set.seed(123)
plist[[1]] <- plot_ordination(physeq=physeq.rel.fam,
                              ordination=ordinate(physeq.rel.fam, "MDS", distance=dist.zhu.fam$UniF))
plist[[2]] <- plot_ordination(physeq=physeq.clr.fam,
                              ordination=ordinate(physeq.clr.fam, "MDS", distance=dist.zhu.fam$Ait))
plist[[3]] <- plot_ordination(physeq=physeq.CSN.fam,
                              ordination=ordinate(physeq.CSN.fam, "MDS", distance=dist.zhu.fam$Bray))
plist[[4]] <- plot_ordination(physeq=physeq.NZcomp.fam,
                              ordination=ordinate(physeq.NZcomp.fam, "MDS", distance=dist.zhu.fam$Canb))
plot.df.fam <- ldply(plist, function(x) x$data)
names(plot.df.fam)[1] <- "distance"

# Plot
ggplot(plot.df.fam, aes(Axis.1, Axis.2, color=host_disease))+
  geom_point(size=4, alpha=0.5) + scale_color_manual(values = c('blue', 'red'))+
  facet_wrap(distance~., scales='free', nrow=1)+
  theme_bw()+
  theme(strip.text.x = element_text(size=20))+
  labs(color="Disease", title="Family agglomeration")
# ggsave(file.path(path.zhu, "distances4_MDS(Family).jpg"), height = 4, width = 15)
```


### 5.3. Plot in 3D
For better visualization, we will also take a glance at reduction to 3D. We will do it on the data agglomerated to Family level (to limit the bias of the ASV sequences).

```{r 3D-function, eval = TRUE, echo = TRUE}
#____________________________________________________________________________________
# Plot 3D ordination
plotDistances3D <- function(d, name_dist){
  
  # Reset parameters
  mds.3D <- NULL
  xyz <- NULL
  fig.3D <- NULL
  
  # Reduce distance matrix to 3 dimensions
  set.seed(123)
  mds.3D <- metaMDS(d, method="MDS", k=3, trace = 0)
  xyz <- scores(mds.3D, display="sites") # pull out the (x,y,z) coordinates
  
  # Plot
  fig.3D <- plot_ly(x=xyz[,1], y=xyz[,2], z=xyz[,3], type="scatter3d", mode="markers",
                    color=sample_data(physeq.zhu.fam)$host_disease, colors = c("blue", "red"))%>%
    layout(title = paste('MDS in 3D with', name_dist, 'distance', sep = ' '))
  
  return(fig.3D)
}
```

Now let's plot!

```{r 3D-plot, fig.width = 8, fig.height = 5}
plotDistances3D(dist.zhu.fam$UniF, "UniFrac")
plotDistances3D(dist.zhu.fam$Ait, "Aitchison")
plotDistances3D(dist.zhu.fam$Canb, "Canberra")
plotDistances3D(dist.zhu.fam$Bray, "Bray-Curtis")
```




****************************
# 6. HIERARCHICAL CLUSTERING
****************************

```{r hierarchical-clust, fig.width = 25, fig.width = 7}
# For heatmaps: have group color
matcol <- data.frame(group = sample_data(physeq.zhu.fam)[,"host_disease"])


# Function to get heatmap from the distances computed
plotHeatmaps <- function(dlist, fontsize){
  
  # Initialize variables
  i=1
  plist <- vector("list", 4)
  names(plist) <- names(dlist)
  
  # Loop through distances
  for(d in dlist){
    plist[[i]] <- pheatmap(as.matrix(d), 
                          clustering_distance_rows = d,
                          clustering_distance_cols = d,
                          fontsize = fontsize,
                          fontsize_col = fontsize-5,
                          fontsize_row = fontsize-5,
                          annotation_col = matcol,
                          annotation_row = matcol,
                          annotation_colors = list(host_disease = c('Healthy' = 'blue', 'IBS' = 'red')),
                          cluster_rows = T,
                          cluster_cols = T,
                          clustering_method = 'complete', # hc method
                          main = names(dlist)[i]) # have name of distance as title
    i <- i+1
  }
  
  return(plist)
}


# Get the heatmaps
heatmp.zhu <- plotHeatmaps(dlist = dist.zhu.fam, fontsize = 8)
```




*******************************
# 7. REPRODUCE PLOTS FROM PAPER
*******************************

Figure 4D - Simpson index of the IBS patients and the healthy controls 

```{r reproduce-fig4, fig.width = 8, fig.height = 5}
# Set order
sample_data(physeq.zhu)$host_disease <- factor(sample_data(physeq.zhu)$host_disease, levels=c("IBS", "Healthy"))

# Plot ACE index
a <- plot_richness(physeq.zhu, x="host_disease", measures="ACE", color="host_disease") +
  geom_boxplot(fill=NA, width=0.3) +
  scale_color_manual(values=c("red", "blue"))+
  geom_signif(comparisons = list(c("Healthy", "IBS")), map_signif_level = FALSE, test="wilcox.test", color="black") +
  theme_cowplot() +
  theme(legend.position = "none")+
  labs(x="", y="ACE")

# Plot Simpson
b <- plot_richness(physeq.zhu, x="host_disease", measures="Simpson", color="host_disease") +
  geom_boxplot(fill=NA, width=0.3) +
  scale_color_manual(values=c("red", "blue"))+
  geom_signif(comparisons = list(c("Healthy", "IBS")), map_signif_level = FALSE, test="wilcox.test", color="black") +
  theme_cowplot() +
  labs(x="", y="Simpson")

ggpubr::ggarrange(a,b,ncol=2, widths=c(0.4,0.6))
```
![Alpha diversity of ace (C) and simpson (D) index of the IBS patients and the healthy controls](/Users/scarcy/Projects/MetaIBS/data/preprocessing/analysis-individual/Zhu-2019/03_EDA-Zhu/fig4.png){width=60%}




******************
# 8. SESSION INFO
******************

```{r session-info}
sessionInfo()
```