---
title: "R Notebook"
output: html_notebook
---

```{r, include=FALSE}
knitr::opts_chunk$set(fig.width = 15, fig.height = 5, warning=FALSE, message=FALSE,
                      root.dir = "/Users/enigma/Desktop/Munich/Praktikum/Data/Rproject_COMBINE/")
```

####################
# IMPORT LIBRARIES #
####################

```{r, echo = FALSE}
library(dada2) # manage fastq file

library(ShortRead) # manage fastq file

library(Biostrings) # to manage dna sequences as strings

library(seqTools) # analyse read length in fastq file

library(phyloseq) # for phyloseq object

library(ggplot2) # for plots
library(ggrepel)
library(reshape2)
library(ggpubr)
library(data.table)

library(plyr)
library(dplyr)

library("doParallel")

library("foreach")

library("DECIPHER") # for phylogenetic tree

library("phangorn")

library("plotly") # plot 3D

library("microbiome") # for centered log-ratio

library("coda") # Aitchison distance

library("coda.base") # Aitchison distance

library("vegan") # NMDS

library(pheatmap) # for heat map

library(RColorBrewer)

library(glmnet)

library(umap)

```


###############
# IMPORT DATA #
###############

```{r import-datasets, echo = TRUE, eval = FALSE}
#________________________________________________________________________OK
# LABUS 2017
physeq.labus <- readRDS("../physeq_labus.rds")
# Check if there are any eukaryota or NA phyla
table(tax_table(physeq.labus)[,1]) # 760 bacteria
table(is.na(tax_table(physeq.labus)[,2])) # 5 NA

physeq.labus <- subset_taxa(physeq.labus, !is.na(Phylum)) # remove the 5 NA Phyla
physeq.labus # sanity check

# Check on metadata
sample_data(physeq.labus)
table(sample_data(physeq.labus)[,'host_subtype']) # IBS-D, C, M, unspecified

# Re-order samples (have Healthy samples then IBS samples)
ordersamples_labus <- rownames(sample_data(physeq.labus)[order(sample_data(physeq.labus)$host_disease),]) # get the order
otu_table(physeq.labus) <- otu_table(physeq.labus)[ordersamples_labus, ] # re-order samples in the phyloseq object
sample_data(physeq.labus) # sanity check

# Keep only samples with >1000 reads
physeq.labus <-  prune_samples(sample_sums(physeq.labus)>=100, physeq.labus)
physeq.labus # 52 samples
#saveRDS(physeq.labus, "../PREDICTION/physeq_labus.rds")


#________________________________________________________________________OK
# LOPRESTI 2019
physeq.lopresti <- readRDS("physeq_lopresti.rds")
# Check if there are any eukaryota or NA phyla
table(tax_table(physeq.lopresti)[,1]) # 1198 bacteria
table(is.na(tax_table(physeq.lopresti)[,2])) # 5 NA at phylum level

physeq.lopresti <- subset_taxa(physeq.lopresti, !is.na(Phylum)) # remove the 5 NA Phyla
physeq.lopresti # sanity check

# Check on metadata
sample_data(physeq.lopresti)
# Re-order samples (have Healthy samples then IBS samples)
ordersamples_lopresti <- rownames(sample_data(physeq.lopresti)[order(sample_data(physeq.lopresti)$host_disease),]) # get the order
otu_table(physeq.lopresti) <- otu_table(physeq.lopresti)[ordersamples_lopresti, ] # re-order samples in the phyloseq object
sample_data(physeq.lopresti) # sanity check

# Keep only samples with >100 reads
physeq.lopresti <- prune_samples(sample_sums(physeq.lopresti)>=100, physeq.lopresti)
physeq.lopresti # 117 samples
#saveRDS(physeq.lopresti, "../PREDICTION/physeq_lopresti.rds")


#________________________________________________________________________OK
# POZUELO 2015
physeq.pozuelo <- readRDS("physeq_pozuelo.rds")
# Check if there are any eukaryota or NA phyla
table(tax_table(physeq.pozuelo)[,1]) # 6222 bacteria
table(is.na(tax_table(physeq.pozuelo)[,1])) # 2 NA at kingdom level

physeq.pozuelo <- subset_taxa(physeq.pozuelo, Kingdom == 'Bacteria' | Kingdom == 'Archaea') # remove the 2 NA Kingdoms and 1 Eukaryota
physeq.pozuelo # sanity check

table(is.na(tax_table(physeq.pozuelo)[,2])) # 105 NA at phylum level
physeq.pozuelo <- subset_taxa(physeq.pozuelo, !is.na(Phylum)) # remove the 105 NA Phyla
physeq.pozuelo # sanity check

# Check on metadata
sample_data(physeq.pozuelo)
# Re-order samples (have Healthy samples then IBS samples)
ordersamples_pozuelo <- rownames(sample_data(physeq.pozuelo)[order(sample_data(physeq.pozuelo)$host_disease),]) # get the order
otu_table(physeq.pozuelo) <- otu_table(physeq.pozuelo)[ordersamples_pozuelo, ] # re-order samples in the phyloseq object
sample_data(physeq.pozuelo)

# Keep only samples with >1000 reads
physeq.pozuelo <- prune_samples(sample_sums(physeq.pozuelo)>=1000, physeq.pozuelo)
physeq.pozuelo # 290 samples
saveRDS(physeq.pozuelo, "../PREDICTION/physeq_pozuelo.rds")

#________________________________________________________________________OK
# ZHUANG 2015
physeq.zhuang <- readRDS("physeq_zhuang.rds")
physeq.zhuang
# Check if there are any eukaryota or NA phyla
table(tax_table(physeq.zhuang)[,1]) # 498 bacteria
table(is.na(tax_table(physeq.zhuang)[,2])) # no NA at phylum level

# Check on metadata
sample_data(physeq.zhuang)
# Re-order samples (have Healthy samples then IBS samples)
ordersamples_zhuang <- rownames(sample_data(physeq.zhuang)[order(sample_data(physeq.zhuang)$host_disease),]) # get the order
otu_table(physeq.zhuang) <- otu_table(physeq.zhuang)[ordersamples_zhuang, ] # re-order samples in the phyloseq object
sample_data(physeq.zhuang)

# Keep only samples with >100 reads
physeq.zhuang <- prune_samples(sample_sums(physeq.zhuang)>=100, physeq.zhuang)
physeq.zhuang # 30 samples
saveRDS(physeq.zhuang, "../PREDICTION/physeq_zhuang.rds")

#________________________________________________________________________OK
# ZEBER-LUBECKA 2016
physeq.zeber <- readRDS("physeq_zeber.rds")
physeq.zeber
# Check if there are any eukaryota or NA phyla
table(tax_table(physeq.zeber)[,1]) # 14,474 bacteria
table(is.na(tax_table(physeq.zeber)[,2])) # 40 NA at phylum level

physeq.zeber <- subset_taxa(physeq.zeber, !is.na(Phylum)) # remove the 40 NA Phyla
physeq.zeber

# Check on metadata
sample_data(physeq.zeber)
sample_data(physeq.zeber)$host_disease <- as.character(sample_data(physeq.zeber)$host_disease)
# Re-order samples (have Healthy samples then IBS samples)
ordersamples_zeber <- rownames(sample_data(physeq.zeber)[order(sample_data(physeq.zeber)$host_disease),]) # get the order
otu_table(physeq.zeber) <- otu_table(physeq.zeber)[ordersamples_zeber, ] # re-order samples in the phyloseq object
sample_data(physeq.zeber)

# Keep only samples with >1000 reads
physeq.zeber <- prune_samples(sample_sums(physeq.zeber)>=1000, physeq.zeber)
physeq.zeber # 90 samples
saveRDS(physeq.zeber, "../PREDICTION/physeq_zeber.rds")

#________________________________________________________________________OK
# ZHU 2019
physeq.zhu <- readRDS("physeq_zhu.rds")
physeq.zhu
# Check if there are any eukaryota or NA phyla
table(tax_table(physeq.zhu)[,1]) # 648 bacteria
table(is.na(tax_table(physeq.zhu)[,2])) # 1 NA at phylum level

physeq.zhu <- subset_taxa(physeq.zhu, !is.na(Phylum)) # remove the 1 NA Phyla
physeq.zhu

# Check on metadata & re-roder (have Healthy samples then IBS samples)
ordersamples_zhu <- rownames(sample_data(physeq.zhu)[order(sample_data(physeq.zhu)$host_disease),]) # get the order 
otu_table(physeq.zhu) <- otu_table(physeq.zhu)[ordersamples_zhu, ] # re-order samples in the phyloseq object
sample_data(physeq.zhu)

# Keep only samples with >1000 reads
physeq.zhu <- prune_samples(sample_sums(physeq.zhu)>=1000, physeq.zhu)
physeq.zhu # 29 samples
#saveRDS(physeq.zhu, "../PREDICTION/physeq_zhu.rds")

#________________________________________________________________________OK
# RINGEL-KULKA 2015
physeq.ringel <- readRDS("physeq_ringel.rds")
physeq.ringel
# Check if there are any eukaryota or NA phyla
table(tax_table(physeq.ringel)[,1]) # 1376 bacteria
table(is.na(tax_table(physeq.ringel)[,2])) # 13 NA at phylum level

physeq.ringel <- subset_taxa(physeq.ringel, !is.na(Phylum)) # remove the 13 NA Phyla
physeq.ringel

# Check on metadata
sample_data(physeq.ringel)

# Keep only samples with >100 reads
physeq.ringel <- prune_samples(sample_sums(physeq.ringel)>=100, physeq.ringel)
physeq.ringel # 75 samples
#saveRDS(physeq.ringel, "../PREDICTION/physeq_ringel.rds")
```



#########################
# CHECK FOR COMMON ASVs #
#########################

```{r common-ASV, echo = TRUE, eval = TRUE}
# Merge phyloseq objects
physeq.zeber
physeq.pozuelo
physeq.zhu

#physeq <- merge_phyloseq(physeq.zeber, physeq.pozuelo)
physeq

physeq_test <- merge_phyloseq(physeq.labus, physeq.lopresti, physeq.pozuelo, physeq.ringel, physeq.zeber, physeq.zhu, physeq.zhuang)
physeq_test <- merge_phyloseq(physeq.ringel, physeq.lopresti)

physeq_test


# Pozuelo and Zhu V4
nb.V4 <- 0

for(i in 1:length(taxa_names(physeq.pozuelo))){
  nb.V4 <- nb.V4+
    length(grep(taxa_names(physeq.pozuelo)[i], taxa_names(physeq.zhu), fixed = TRUE)) # get nb of times the OTU is found in zhu
}

nb.V4

mean(length)

table(taxa_names(physeq.pozuelo) %in% taxa_names(physeq.zhu))

# Look at length taxa_names
mean(nchar(taxa_names(physeq.zhu)))
sd(nchar(taxa_names(physeq.zhu)))
mean(nchar(taxa_names(physeq.pozuelo)))
sd(nchar(taxa_names(physeq.pozuelo)))

median(nchar(taxa_names(physeq.zhu)))
median(nchar(taxa_names(physeq.pozuelo)))

taxa_names(physeq.zhu)[1:2]
substr(taxa_names(physeq.zhu)[1:2], start = 1, stop = 10)

for (i in 1:420){
  table(substr(taxa_names(physeq.zhu), start = i, stop = i+248) %in% taxa_names(physeq.pozuelo))
}

table(substr(taxa_names(physeq.zhu), start = 160, stop = 413) %in% taxa_names(physeq.pozuelo))

# 9 OTUs of length 246bp are common
# pozuelo: 200 - 248 bp

sample_data(physeq_test)

```


#########
# MERGE #
#########

# MERGE & PREPROCESS PHYLOSEQ OBJECT
```{r}
# MERGE
physeq <- merge_phyloseq(physeq.ringel, physeq.labus, physeq.lopresti, physeq.pozuelo, physeq.zhuang, physeq.zhu, physeq.zeber)
physeq
#table(tax_table(physeq)[,1]) # 13 archeae
sample_data(physeq)[is.na(sample_data(physeq)[,'host_sex']), 'host_sex'] <- 'unknown'
sample_data(physeq)[is.na(sample_data(physeq)$sample_type), 'sample_type'] <- 'stool'

# check on sample_data
sample_data(physeq)
table(sample_data(physeq)$author) #ok
table(sample_data(physeq)$sequencing_tech) #ok
table(sample_data(physeq)$variable_region) # ok
table(sample_data(physeq)$host_disease) # ok
table(sample_data(physeq)$host_sex) # ok

table(sample_data(physeq)$host_age) # ok
table(sample_data(physeq)$host_subtype) # ok
table(sample_data(physeq)$sample_type) # ok
table(sample_data(physeq)$time_point) # ok

# Save for prediction
saveRDS(as.data.frame(sample_data(physeq)), "../PREDICTION/metadata.rds")

# REMOVE SAMPLES WITH LESS THAN 1000 TOTAL COUNTS
physeq_best <-  prune_samples(sample_sums(physeq)>=1000, physeq)
physeq_best # 616 samples

# KEEP ONLY FECAL SAMPLES
physeq_best.fecal <- subset_samples(physeq_best, sample_type == 'stool')
physeq_best.fecal # 605 samples

#saveRDS(physeq_best.fecal, "physeq_best_fecal.rds")

# NORMALIZE TOTAL COUNT PER SAMPLE TO 1
physeq.rel <- physeq_best.fecal
physeq.rel <- transform_sample_counts(physeq.rel, function(x) x / sum(x) ) # divide each count by the total number of counts (per sample)
# sanity check
sum(otu_table(physeq.rel) < 0) # see how many negative values are present in the matrix
table(rowSums(otu_table(physeq.rel))) # check if there is any row not summing to 1
physeq.rel

# ADD PSEUDOCOUNTS
physeq.pseudocts <- physeq.rel
otu_table(physeq.pseudocts)[otu_table(physeq.pseudocts) == 0] <- 3e-6
# check the 0 values have been replaced
otu_table(physeq.rel)[1:2,1:2]
otu_table(physeq.pseudocts)[1:2,1:2]


# Plot heatmap
physeq.rel.top <- prune_taxa(names(sort(taxa_sums(physeq.pseudocts),TRUE)[1:100]), physeq.pseudocts) # get the top 100 most abundant OTUs

# How many phyla?
length(table(tax_table(physeq)[, "Phylum"]))
length(table(tax_table(physeq)[, "Family"]))
table(is.na(tax_table(physeq)[, "Family"]))
table(is.na(tax_table(physeq)[, "Genus"]))
table(is.na(tax_table(physeq)[, "Species"]))
length(table(tax_table(physeq)[, "Species"]))
median(c(755,1193,6114,498,14434,647,1363))
```

```{r TO-DELETE}
# Relative abundance for Phylum
phylumTable <- physeq_best.fecal %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

mean(phylumTable[phylumTable$Phylum == "Firmicutes" ,"Abundance"])+mean(phylumTable[phylumTable$Phylum == "Bacteroidota" ,"Abundance"])

mean(phylumTable[phylumTable$Phylum == "Proteobacteria" ,"Abundance"])+
mean(phylumTable[phylumTable$Phylum == "Verrucomicrobiota" ,"Abundance"])+
mean(phylumTable[phylumTable$Phylum == "Actinobacteriota" ,"Abundance"])+
mean(phylumTable[phylumTable$Phylum == "Fusobacteriota" ,"Abundance"])

jpeg("./Figures/phylum_legend.jpg", height = 10000, width = 5000, res = 400)
ggplot(phylumTable.fecal,
       aes(x = reorder(Sample, Sample, function(x) mean(phylumTable.fecal[Sample == x & Phylum == 'Bacteroidota', 'Abundance'])),
           y = Abundance, fill = Phylum, width = 0.8))+
  facet_grid(author~host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  scale_fill_manual(values = phyla.col)+
  theme_classic()+
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.line.x = element_blank(),
        axis.title.y = element_text(size = 40),
        legend.text = element_text(size = 30),
        legend.title = element_text(size = 30))+
  labs(x = "", y = "Relative abundance")+
  scale_y_continuous(expand = c(0, 0))
dev.off()
```


# HEATMAP

```{r heatmap-top-ASVs}
# Create categories
mat_col <- data.frame(disease = sample_data(physeq.rel.top)[,'host_disease'],
                      seq_tech = sample_data(physeq.rel.top)[,'sequencing_tech'],
                      author = sample_data(physeq.rel.top)[,'author'])
mat_col$author <- factor(mat_col$author, levels = c('Ringel-Kulka', 'Labus', "LoPresti",
                                                       "Pozuelo", "Zhuang", "Zhu", "Zeber-Lubecka"))
#saveRDS(mat_col, 'mat_col.rds')
table(mat_col$author)
#mat_col <- data.frame(disease = sample_data(physeq.rel.top)[,'host_disease'])
#mat_col$host_disease <- as.factor(mat_col$host_disease)
#mat_col$sequencing_tech <- as.factor(mat_col$sequencing_tech)

annotationCol <- list(host_disease = c(Healthy = 'blue', IBS = 'red'),
                                  sequencing_tech = c('454 pyrosequencing' = '#6600FF', 'Illumina MiSeq' = '#00CC66', 'Ion Torrent' = '#FF6633'),
                                  author = setNames(brewer.pal(nlevels(mat_col$author), "Spectral"), levels(mat_col$author)))

#valueCol <- colorRampPalette(c(rev(brewer.pal(9, "Blues")[1:3]), rev(brewer.pal(9, "RdBu")[1:4])), bias = 0.7)(100)
valueCol <- colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100)

# Plot heatmap
jpeg("./Plots/Heatmaps/plot_heatmap_top100OTU_06.jpg", width = 2000, height = 1600, res = 200)
pheatmap(log10(t(as.matrix(otu_table(physeq.rel.top)))),
         color = valueCol,
         show_rownames = F,
         show_colnames = F,
         fontsize_col = 1,
         cluster_rows = T,
         cluster_cols = F,
         #cutree_rows = 5,
         #na_col = "grey",
         clustering_method = 'ward.D',
         annotation_col = mat_col,
         annotation_colors = annotationCol,
         main = "Hierarchical clustering with ward linkage (top 100 most abundant OTUs)")
dev.off()

# Save the samples ordering
samples_order <- rownames(sample_data(physeq.rel.top))

```


```{r heatmap-recolor}
quantile_breaks <- function(xs, n) {
  print('Hello')
  breaks <- quantile(xs, probs = seq(0, 1, length.out = n))
  print(breaks)
  breaks[!duplicated(breaks)]
}

mat_breaks <- quantile_breaks(xs = t(otu_table(physeq.rel.top)), n = 10)

quantile(otu_table(physeq.rel.top), probs = seq(0, 1, length.out = 10))

pheatmap(log10(t(as.matrix(otu_table(physeq.rel.top)))),
         show_rownames = F,
         show_colnames = F,
         fontsize_col = 1,
         cluster_rows = T,
         cluster_cols = F,
         clustering_method = 'ward.D',
         annotation_col = mat_col,
         annotation_colors = list(host_disease = c(Healthy = 'blue', IBS = 'red'),
                                  sequencing_tech = c('454 pyrosequencing' = '#6600FF', 'Illumina MiSeq' = '#00CC66', 'Ion Torrent' = '#FF6633'),
                                  author = setNames(brewer.pal(nlevels(mat_col$author), "Spectral"), levels(mat_col$author))),
         main = "Hierarchical clustering with ward linkage (top 100 most abundant OTUs)")
```


#############
# AGGREGATE #
#############
```{r aggregate-genus}
#physeq.rel.genus <- tax_glom(physeq.rel, "Genus")
test.physeq <- prune_taxa(names(sort(taxa_sums(physeq.fecal),TRUE)[1:1000]), physeq.fecal)
genus.agg <- physeq.fecal %>%
  tax_glom(taxrank = "Genus") %>%                     # agglomerate at genus level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

# Get dataframe
genusTable <- acast(genus.agg, Genus ~ Sample, value.var = 'Abundance')
genusTable <- genusTable[,samples_order] # reorder samples
genusTable[genusTable == 0] <- 10e-7 # add pseudocounts for heatmap
#genus.agg[genus.agg$Genus == 'Alistipes' & genus.agg$Sample == 'ERR1051013','Abundance'] # sanity check
table(colSums(genusTable)) # sanity check (sum per sample == 1)
table(rownames(genusTable))

jpeg("./New_Figures/plot_heatmap_genera.jpg", width = 8000, height = 6000, res = 800)
pheatmap(log10(genusTable),
         color =  colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
         show_rownames = T,
         show_colnames = F,
         fontsize_row = 4,
         cluster_rows = T,
         cluster_cols = T,
         cutree_cols = 2,
         clustering_method = 'ward.D',
         annotation_col = mat_col,
         annotation_colors = annotationCol,
         main = "Hierarchical clustering with ward linkage (genera as rows, top 1000 ASVs)")
dev.off()
```


```{r aggregate-family}
#physeq.rel.family <- tax_glom(physeq.rel, "Family")

fam.agg <- physeq_best.fecal %>%
  tax_glom(taxrank = "Family") %>%                     # agglomerate at genus level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

# Get dataframe
familyTable <- acast(fam.agg, Family ~ Sample, value.var = 'Abundance')
familyTable <- familyTable[,samples_order] # reorder samples
familyTable[familyTable == 0] <- 10e-7 # add pseudocounts for heatmap
#fam.agg[fam.agg$Family == 'Bacillaceae' & fam.agg$Sample == 'ERR1051013','Abundance'] # sanity check
table(colSums(familyTable)) # sanity check (sum per sample == 1)
table(rownames(familyTable))

jpeg("./Plots/Heatmaps/plot_heatmap_family_04.jpg", width = 4000, height = 3000, res = 400)
pheatmap(log10(familyTable),
         color = valueCol,
         show_rownames = T,
         show_colnames = F,
         fontsize_row = 3,
         cluster_rows = T,
         cluster_cols = T,
         cutree_cols = 2,
         clustering_method = 'ward.D',
         annotation_col = mat_col,
         annotation_colors = annotationCol,
         main = "Hierarchical clustering with ward linkage (families as rows)")
dev.off()
```


```{r aggregate-order}
#physeq.rel.order <- tax_glom(physeq.rel, "Order")

order.agg <- physeq_best.fecal %>%
  tax_glom(taxrank = "Order") %>%                     # agglomerate at genus level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

# Get dataframe
orderTable <- acast(order.agg, Order ~ Sample, value.var = 'Abundance')
orderTable <- orderTable[,samples_order] # reorder samples
orderTable[orderTable == 0] <- 10e-7 # add pseudocounts for heatmap
#order.agg[order.agg$Order == 'Bacillales' & order.agg$Sample == 'ERR1051013','Abundance'] # sanity check
table(colSums(orderTable)) # sanity check (sum per sample == 1)
table(rownames(orderTable))

jpeg("./Plots/Heatmaps/plot_heatmap_order_02.jpg", width = 4000, height = 3000, res = 400)
pheatmap(log10(orderTable),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
         show_rownames = T,
         show_colnames = F,
         fontsize_row = 5,
         cluster_rows = T,
         cluster_cols = F,
         clustering_method = 'ward.D',
         annotation_col = mat_col,
         annotation_colors = list(host_disease = c(Healthy = 'blue', IBS = 'red'),
                                  sequencing_tech = c('454 pyrosequencing' = '#6600FF',
                                                      'Illumina MiSeq' = '#00CC66',
                                                      'Ion Torrent' = '#FF6633'),
                                  author = setNames(brewer.pal(nlevels(mat_col$author), "Spectral"), levels(mat_col$author))),
         main = "Hierarchical clustering with ward linkage (orders as rows)")
dev.off()
```


```{r aggregate-class}
#physeq.rel.class <- tax_glom(physeq.rel, "Class")

class.agg <- physeq_best.fecal %>%
  tax_glom(taxrank = "Class") %>%                     # agglomerate at genus level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

# Get dataframe
classTable <- acast(class.agg, Class ~ Sample, value.var = 'Abundance')
classTable <- classTable[,samples_order] # reorder samples
classTable[classTable == 0] <- 10e-7 # add pseudocounts for heatmap
#class.agg[class.agg$Order == 'Bacilli' & class.agg$Sample == 'ERR1051013','Abundance'] # sanity check
table(colSums(classTable)) # sanity check (sum per sample == 1)
table(rownames(classTable))

jpeg("./Plots/Heatmaps/plot_heatmap_class_02.jpg", width = 4000, height = 3000, res = 400)
pheatmap(log10(classTable),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
         show_rownames = T,
         show_colnames = F,
         fontsize_row = 7,
         cluster_rows = T,
         cluster_cols = F,
         clustering_method = 'ward.D',
         annotation_col = mat_col,
         annotation_colors = list(host_disease = c(Healthy = 'blue', IBS = 'red'),
                                  sequencing_tech = c('454 pyrosequencing' = '#6600FF', 'Illumina MiSeq' = '#00CC66', 'Ion Torrent' = '#FF6633'),
                                  author = setNames(brewer.pal(nlevels(mat_col$author), "Spectral"), levels(mat_col$author))),
         main = "Hierarchical clustering with ward linkage (classes as rows)")
dev.off()
```



```{r aggregate-phylum}
#physeq.rel.phylum <- tax_glom(physeq.rel, "Phylum")
phylum.agg <- physeq.fecal %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at genus level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

# Get dataframe
phylumTable <- acast(phylum.agg, Phylum ~ Sample, value.var = 'Abundance')
phylumTable <- phylumTable[,samples_order] # reorder samples
phylumTable[phylumTable == 0] <- 10e-6
#phylum.agg[phylum.agg$Order == 'Bacilli' & phylum.agg$Sample == 'ERR1051013','Abundance'] # sanity check
table(colSums(phylumTable)) # sanity check (sum per sample == 1)
table(rownames(phylumTable))
table(rowSums(phylumTable) == 0)


# Create categories
mat_col <- data.frame(disease = sample_data(physeq.fecal)[,'host_disease'],
                      seq_tech = sample_data(physeq.fecal)[,'sequencing_tech'],
                      author = sample_data(physeq.fecal)[,'author'])
mat_col$author <- factor(mat_col$author, levels = c('Labus', "LoPresti", "Pozuelo",
                                                    "Zhuang", "Zhu", 'Hugerth',
                                                    "Zeber-Lubecka", 'Nagel'))
table(mat_col$author)
#mat_col <- data.frame(disease = sample_data(physeq.rel.top)[,'host_disease'])
#mat_col$host_disease <- as.factor(mat_col$host_disease)
#mat_col$sequencing_tech <- as.factor(mat_col$sequencing_tech)

annotationCol <- list(host_disease = c(Healthy = 'blue', IBS = 'red'),
                      sequencing_tech = c('454 pyrosequencing' = '#6600FF',
                                          'Illumina' = '#00CC66',
                                          'Ion Torrent' = '#FF6633'),
                      author = setNames(brewer.pal(nlevels(mat_col$author), "Spectral"),
                                        levels(mat_col$author)))


jpeg("./New_Figures/plot_heatmap_phyla.jpg", width = 4000, height = 3000, res = 400)
pheatmap(log10(phylumTable),
         color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
         show_rownames = T,
         show_colnames = F,
         fontsize_row = 8,
         cluster_rows = T,
         cluster_cols = T,
         #cutree_rows = 2,
         cutree_cols = 2,
         clustering_method = 'ward.D',
         annotation_col = mat_col,
         annotation_colors = annotationCol,
         main = "Hierarchical clustering with ward linkage (phyla as rows)")
dev.off()
```



################
### DISTANCES ##
################

```{r functions-distances}
#____________________________________________________________________________________
# Measure distances
Distances <- function(physeq_rel, physeq_CSN, physeq_clr){
  print("Aitchison")
  glom.ait <- dist(x = otu_table(physeq_clr), method = 'euc') # aitchison
  print("Bray")
  glom.bray <- phyloseq::distance(physeq_CSN, method = "bray") # bray-curtis
  print("Canberra")
  glom.can <- phyloseq::distance(physeq_rel, method = "canberra") # canberra
  
  dist.list <- list("Ait" = glom.ait, "Canb" = glom.can, "Bray" = glom.bray)
  return(dist.list)
}


#____________________________________________________________________________________
# Plot 2D ordination
MDS_2D_disease <- function(physeq_rel, physeq_CSN, physeq_clr, dist){
  
  plist <- NULL
  plist <- vector("list", 3) # save each plot to a list
  names(plist) <- c("Aitchison", "Bray-Curtis", "Canberra") # save the name of each method
  
  # Aitchison
  print('Aitchison')
  iMDS.ait  <- ordinate(physeq_clr, "MDS", distance=dist$Ait)
  p <- NULL
  p <- plot_ordination(physeq_clr, iMDS.ait, color="host_disease") +
    ggtitle("MDS using distance method Aitchison")
  plist[[1]] <- p
  
  # Bray-Curtis
  print('Bray')
  iMDS.bray  <- ordinate(physeq_CSN, "MDS", distance=dist$Bray)
  p <- NULL
  p <- plot_ordination(physeq_CSN, iMDS.bray, color="host_disease") +
    ggtitle("MDS using distance method Bray-Curtis")
  plist[[2]] <- p
  
  # Canberra
  print('Canb')
  iMDS.canb  <- ordinate(physeq_rel, "MDS", distance=dist$Canb)
  p <- NULL
  p <- plot_ordination(physeq_rel, iMDS.canb, color="host_disease") +
    ggtitle("MDS using distance method Bray-Curtis")
  plist[[3]] <- p
  
  # Creating a dataframe to plot everything
  print('Extracting data')
  plot.df = ldply(plist, function(x) x$data)
  names(plot.df)[1] <- "distance"
  plot.df[is.na(plot.df$host_disease), "host_disease"] <- "NA" # for Ringel samples
  
  # Plot
  print('Plotting')
  p.alldist <-  ggplot(plot.df, aes(Axis.1, Axis.2, color=host_disease))+
    geom_point(size=10, alpha=0.7)+
    facet_wrap(~distance, scales='free')

  return(p.alldist)
}

MDS_2D_author <- function(physeq_rel, physeq_CSN, physeq_clr, dist){
  
  plist <- NULL
  plist <- vector("list", 3) # save each plot to a list
  names(plist) <- c("Aitchison", "Bray-Curtis", "Canberra") # save the name of each method
  
  # Aitchison
  print('Aitchison')
  iMDS.ait  <- ordinate(physeq_clr, "MDS", distance=dist$Ait)
  p <- NULL
  p <- plot_ordination(physeq_clr, iMDS.ait, color="author") +
    ggtitle("MDS using distance method Aitchison")
  plist[[1]] <- p
  
  # Bray-Curtis
  print('Bray')
  iMDS.bray  <- ordinate(physeq_CSN, "MDS", distance=dist$Bray)
  p <- NULL
  p <- plot_ordination(physeq_CSN, iMDS.bray, color="author") +
    ggtitle("MDS using distance method Bray-Curtis")
  plist[[2]] <- p
  
  # Canberra
  print('Canb')
  iMDS.canb  <- ordinate(physeq_rel, "MDS", distance=dist$Canb)
  p <- NULL
  p <- plot_ordination(physeq_rel, iMDS.canb, color="author") +
    ggtitle("MDS using distance method Bray-Curtis")
  plist[[3]] <- p
  
  # Creating a dataframe to plot everything
  print('Extracting data')
  plot.df = ldply(plist, function(x) x$data)
  names(plot.df)[1] <- "distance"
  plot.df[plot.df$author == "LoPresti", "author"] <- "Lo Presti" # adapt author name
  plot.df$author <- factor(plot.df$author, levels = authors_order) # put authors in the right order
  
  # Plot
  print('Plotting')
  p.alldist <-  ggplot(plot.df, aes(Axis.1, Axis.2, color=author))+
    geom_point(size=10, alpha=0.7)+
    facet_wrap(~distance, scales='free')

  return(p.alldist)
}

MDS_2D_seqTech <- function(physeq_rel, physeq_CSN, physeq_clr, dist){
  
  plist <- NULL
  plist <- vector("list", 3) # save each plot to a list
  names(plist) <- c("Aitchison", "Bray-Curtis", "Canberra") # save the name of each method
  
  # Aitchison
  print('Aitchison')
  iMDS.ait  <- ordinate(physeq_clr, "MDS", distance=dist$Ait)
  p <- NULL
  p <- plot_ordination(physeq_clr, iMDS.ait, color="sequencing_tech") +
    ggtitle("MDS using distance method Aitchison")
  plist[[1]] <- p
  
  # Bray-Curtis
  print('Bray')
  iMDS.bray  <- ordinate(physeq_CSN, "MDS", distance=dist$Bray)
  p <- NULL
  p <- plot_ordination(physeq_CSN, iMDS.bray, color="sequencing_tech") +
    ggtitle("MDS using distance method Bray-Curtis")
  plist[[2]] <- p
  
  # Canberra
  print('Canb')
  iMDS.canb  <- ordinate(physeq_rel, "MDS", distance=dist$Canb)
  p <- NULL
  p <- plot_ordination(physeq_rel, iMDS.canb, color="sequencing_tech") +
    ggtitle("MDS using distance method Bray-Curtis")
  plist[[3]] <- p
  
  # Creating a dataframe to plot everything
  print('Extracting data')
  plot.df = ldply(plist, function(x) x$data)
  names(plot.df)[1] <- "distance"
  
  # Plot
  print('Plotting')
  p.alldist <-  ggplot(plot.df, aes(Axis.1, Axis.2, color=sequencing_tech))+
    geom_point(size=10, alpha=0.5)+
    facet_wrap(~distance, scales='free')

  return(p.alldist)
}

#____________________________________________________________________________________
# Plot 3D ordination
MDS_3D <- function(d, name_dist, physeq_obj){
  
  # Reset parameters
  mds.3D <- NULL
  xyz <- NULL
  fig.3D <- NULL
  
  # Reduce distance matrix to 3 dimensions
  print("Reducing to 3D")
  set.seed(123) # to get the same dimensionality reduction at each run
  mds.3D <- metaMDS(d, method="MDS", k=3, trace = 1)
  print("Pulling out xyz")
  xyz <- scores(mds.3D, display="sites") # pull out the x y z coordinates
  
  print("Plotting figure")
  fig.3D <- plot_ly(x=xyz[,1], y=xyz[,2], z=xyz[,3], type="scatter3d", mode="markers",
                    color=sample_data(physeq_obj)$author)%>%
    #add_trace(x=xyz[,1], y=xyz[,2], z=xyz[,3],
    #       type='scatter3d', mode='text', text = rownames(xyz), textfont = list(color = "black")) %>%
    layout(title = paste('MDS in 3D with', name_dist, 'distance', sep = ' '))
  
  return(fig.3D)
}
```


```{r}
# BRAY ISSUE
test <- phyloseq::distance(physeq.CSN, method = "bray") # bray-curtis
table(as.matrix(test) == 0)
testMDS <- ordinate(physeq.CSN, "MDS", distance="bray")
plot_ordination(physeq.CSN, testMDS, color="author")
```

```{r}
test[is.na(test$host_disease), "host_disease"] <- "NA"
test[test$author == "LoPresti", "author"] <- "Lo Presti"
table(test$author)
authors_order
test$author <- factor(test$author, levels = authors_order)
```



```{r dist-noglom, fig.height = 5, fig.width = 10}
# NORMALIZATION
physeq.clr <- physeq.fecal
physeq.clr <- microbiome::transform(physeq.clr, "clr")

physeq.CSN <- physeq.fecal
otu_table(physeq.CSN)[otu_table(physeq.CSN) == 0] <- 0.5
min(sample_sums(physeq.fecal))
physeq.CSN <- transform_sample_counts(physeq.CSN, function(x) (x*529) / sum(x) ) # sample_sum = min seq depth
#physeq.CSN # sanity check
#table(sample_sums(physeq.CSN)) # sanity check

physeq.NZ <- physeq.fecal
otu_table(physeq.NZ)[otu_table(physeq.NZ) == 0] <- 0.5
physeq.NZ <- transform_sample_counts(physeq.NZ, function(x) x / sum(x) ) # non-zero compositions
#table(sample_sums(physeq.NZ)) # sanity check

# COMPUTE DISTANCES
dist.no_glom <- Distances(physeq_rel = physeq.NZ, physeq_clr = physeq.clr, physeq_CSN = physeq.CSN)

#__________________________________
# PLOTTING 2D
alldist.2D.no_glom <- MDS_2D_disease(physeq_rel = physeq.NZ, physeq_clr = physeq.clr,
                                     physeq_CSN = physeq.CSN, dist = dist.no_glom)
p1 <- alldist.2D.no_glom +
  scale_color_manual(values = c('blue', 'red'), labels = c("Zeber-Lubecka", "IBS"))+
  theme_bw()+
  theme(strip.text = element_text(size = 70, margin = margin(1,1,1,1, "cm")),
        strip.background = element_rect(fill = "white", size = 3),
        legend.text = element_text(size = 40),
        legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(color = "Disease status")


author.2D.no_glom <- MDS_2D_author(physeq_rel = physeq.NZ, physeq_clr = physeq.clr,
                                   physeq_CSN = physeq.CSN, dist = dist.no_glom)
p2 <- author.2D.no_glom+
  theme_bw()+
  theme(strip.text = element_blank(),
        #strip.background = element_rect(fill = "white", size = 2),
        legend.text = element_text(size = 40),
        legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(color = "Datasets")


seqTech.2D.no_glom <- MDS_2D_seqTech(physeq_rel = physeq.NZ, physeq_clr = physeq.clr,
                                   physeq_CSN = physeq.clr, dist = dist.no_glom)
seqTech.2D.no_glom
p3 <- seqTech.2D.no_glom+
  theme_bw()+
  theme(strip.text = element_blank(),
        #strip.background = element_rect(fill = "white", size = 2),
        legend.text = element_text(size = 40),
        legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(color = "Sequencing Technology")

#__________________________________
# PLOTTING 3D
Ait.3D.no_glom <- MDS_3D(d = dist.no_glom$Ait, name_dist = 'Canberra', physeq_obj = physeq.clr)
Bray.3D.no_glom <- MDS_3D(d = dist.no_glom$Bray, name_dist = 'Bray-Curtis', physeq_obj = physeq.CSN)
Bray.3D.no_glom.author <- MDS_3D(d = dist.no_glom$Bray, name_dist = 'Bray-Curtis', physeq_obj = physeq.CSN)
Canb.3D.no_glom <- MDS_3D(d = dist.no_glom$Canb, name_dist = 'Canberra', physeq_obj = physeq.NZ)

Ait.3D.no_glom
Bray.3D.no_glom
Bray.3D.no_glom.author
Canb.3D.no_glom
```


```{r}
# Bray physeq.CSN
# Canberra physeq.NZ
ordu = ordinate(physeq.CSN, "PCoA", "bray")

plot.age <- plot_ordination(subset_samples(physeq.CSN, host_age > 0),
                ordinate(subset_samples(physeq.CSN, host_age > 0), "PCoA", "bray"),
                color="host_age")+
  scale_colour_gradient2(low = muted("red"), mid = "white", high = muted("blue"), midpoint = 45)+
  theme_classic()+
  labs(color = "Age")
jpeg("./New_Figures/PCA_age.jpg", width = 2000, height = 1500, res = 400)
plot.age
dev.off()
```

```{r}
test.physeq <- subset_samples(physeq, author %in% c("LoPresti", "Hugerth"))
otu_table(test.physeq)[otu_table(test.physeq) == 0] <- 0.5
min(sample_sums(test.physeq))
test.physeq <- transform_sample_counts(test.physeq, function(x) (x*517) / sum(x) ) # sample_sum = min seq depth

y <- plot_ordination(subset_samples(test.physeq, author=="Hugerth"),
                ordinate(subset_samples(test.physeq, author=="Hugerth"), "PCoA", "bray"))
plotdf =  y$data

jpeg("./New_Figures/PCA_sampletype.jpg", width = 2000, height = 1500, res = 350)
ggplot(plotdf, aes(Axis.1, Axis.2, color=sample_type))+
  geom_line(aes(group = host_ID, alpha=0.3), color = "grey", lwd = 0.5)+
  geom_point(size=2, alpha=0.8)+
  scale_color_manual(values=c("sigmoid" = "purple", "stool" = "#66CC00"))+
  theme_classic()+
  labs(color = "Sample type")
dev.off()



plot_ordination(subset_samples(test.physeq, author=="Hugerth"),
                ordinate(subset_samples(test.physeq, author=="Hugerth"), "PCoA", "bray"),
                color="sample_type")+
  geom_line(aes(group = host_ID, alpha=0.3), color = "grey", lwd = 0.5)+
  scale_color_manual(values = c("sigmoid" = "purple", "stool" = "#33FF00"))+
  theme_classic()+
  labs(color = "Sample type")
```


```{r}
# Bray physeq.CSN
# Canberra physeq.NZ
ordu = ordinate(physeq.CSN, "PCoA", "bray")
plot.bmi <- plot_ordination(subset_samples(physeq.CSN, host_bmi > 0),
                ordinate(subset_samples(physeq.CSN, host_bmi > 0), "PCoA", "bray"),
                color="host_bmi")+
  scale_colour_gradient2(low = muted("red"), mid = "white", high = muted("blue"), midpoint = 26)+
  theme_classic()+
  labs(color = "BMI")

jpeg("./New_Figures/PCA_bmi.jpg", width = 2000, height = 1500, res = 400)
plot.bmi
dev.off()

plot.sex <- plot_ordination(subset_samples(physeq.CSN, host_sex %in% c("male", "female")),
                ordinate(subset_samples(physeq.CSN, host_sex %in% c("male", "female")), "PCoA", "bray"),
                color="host_sex")+
  scale_color_manual(values=c("female" = '#FF9999', "male" = "#99CCFF"))+
  theme_classic()+
  labs(color = "Gender")

jpeg("./New_Figures/PCA_gender.jpg", width = 2000, height = 1500, res = 350)
plot.sex
dev.off()
```

```{r}

plot_ordination(physeq.CSN, ordinate(physeq.CSN, "PCoA", distance=dist.no_glom$Bray))
plot.df =  x$data

jpeg("./New_Figures/PCA_disease.jpg", width = 2000, height = 1500, res = 350)
ggplot(plot.df, aes(Axis.1, Axis.2, color=host_disease))+
    geom_point(size=3, alpha=0.5)+
  scale_color_manual(values=c("Healthy" = 'blue', "IBS" = "red"))+
  theme_classic()+
  labs(color = "Disease")
dev.off()

plot.df$author <- factor(plot.df$author, levels=c("Labus", "LoPresti", "Pozuelo","Zhuang", "Zhu", "Hugerth", "Zeber-Lubecka", "Nagel"))

jpeg("./New_Figures/PCA_authors.jpg", width = 2000, height = 1500, res = 350)
ggplot(plot.df, aes(Axis.1, Axis.2, color=author))+
    geom_point(size=2, alpha=0.5)+
  theme_classic()+
  labs(color = "Datasets")
dev.off()

jpeg("./New_Figures/PCA_seqtech.jpg", width = 2300, height = 1500, res = 350)
ggplot(plot.df, aes(Axis.1, Axis.2, color=sequencing_tech))+
    geom_point(size=3, alpha=0.5)+
  scale_color_manual(values=c("454 pyrosequencing" = 'purple',
                              "Illumina" = "#66CC00",
                              "Ion Torrent" = "#FF6600"))+
  theme_classic()+
  labs(color = "Sequencing technology")
dev.off()
```


```{r export-plot-noglom, eval = FALSE, echo = FALSE}
library(ggpubr)

jpeg("./New_Figures/MDS_mergedData_02.jpg", width = 2700, height = 3000, res = 90)
ggarrange(p1, p2, p3, nrow = 3)
dev.off()

jpeg("./Plots/AllDistances_MDS.jpg", width = 3000, height = 1000, res = 90)
alldist.2D.no_glom +
  scale_color_manual(values = c('blue', 'red'))+
  theme(axis.text.x = element_text(size = 20),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 30),
        title = element_text(size = 30),
        strip.text = element_text(size = 50))+
  labs(title = "No agglomeration")
dev.off()

jpeg("./Plots/AllDistances_Datasets.jpg", width = 3000, height = 1000, res = 90)
dataset.2D.no_glom +
  theme(axis.text.x = element_text(size = 20),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 30),
        title = element_text(size = 30),
        strip.text = element_text(size = 50))+
  labs(title = "No agglomeration")
dev.off()
```


```{r aitchison-check, eval = FALSE, echo = FALSE}
test.ait <- dist(x = otu_table(physeq.rel), method = 'ait') # aitchison
test.ait <- as.matrix(test.ait)
rownames(test.ait) <- rownames(as.matrix(dist.no_glom$Ait))
colnames(test.ait) <- colnames(as.matrix(dist.no_glom$Ait))

iMDS.test  <- ordinate(physeq.rel, "MDS", distance=as.dist(test.ait))
plot_ordination(physeq.rel, iMDS.test, color="host_disease") +
    ggtitle("MDS using distance method Aitchison")+ scale_color_manual(values = c('blue', 'red'))

```



```{r dist-phyl-glom, fig.height = 5, fig.width = 10}
# Taxonomic agglomeration
phylumGlom.physeq <- tax_glom(physeq_best.fecal, "Phylum")

# NORMALIZATION
phylumGlom.pseudocts <- phylumGlom.physeq # pseudocounts
otu_table(phylumGlom.pseudocts)[otu_table(phylumGlom.pseudocts) == 0] <- 0.5
phylumGlom.rel <- phylumGlom.pseudocts # compositional
phylumGlom.rel <- transform_sample_counts(phylumGlom.rel, function(x) x / sum(x) )
phylumGlom.clr <- phylumGlom.pseudocts # CLR transformation
phylumGlom.clr <- microbiome::transform(phylumGlom.clr, "clr")
phylumGlom.CSN <- phylumGlom.physeq # common-scale normalization
phylumGlom.CSN <- transform_sample_counts(phylumGlom.CSN, function(x) x / sum(x) )

# COMPUTE DISTANCES
dist.phylumGlom <- Distances(physeq_rel = phylumGlom.rel, physeq_clr = phylumGlom.clr, physeq_CSN = phylumGlom.CSN)

# PLOTTING 2D
alldist.2D.phylumGlom <- MDS_2D(physeq_rel = phylumGlom.rel, physeq_clr = phylumGlom.clr,
                                physeq_CSN = phylumGlom.CSN, dist = dist.phylumGlom)
alldist.2D.phylumGlom + scale_color_manual(values = c('blue', 'red')) + labs(title = "Agglomeration at Phylum level")

dataset.2D.phylumGlom <- MDS_2D_Dataset(physeq_rel = phylumGlom.rel, physeq_clr = phylumGlom.clr,
                                        physeq_CSN = phylumGlom.CSN, dist = dist.phylumGlom)
dataset.2D.phylumGlom + labs(title = "Agglomeration at Phylum level")

# PLOTTING 3D
Ait.3D.no_glom <- MDS_3D(d = dist.no_glom$Ait, name_dist = 'Canberra', physeq_obj = physeq.clr)
Bray.3D.no_glom <- MDS_3D(d = dist.no_glom$Bray, name_dist = 'Bray-Curtis', physeq_obj = physeq.CSN)
Canb.3D.no_glom <- MDS_3D(d = dist.no_glom$Canb, name_dist = 'Canberra', physeq_obj = physeq.rel)

Ait.3D.no_glom
Bray.3D.no_glom
Canb.3D.no_glom
```

```{r export-plot-phylumGlom}
jpeg("./Plots/phylumGlom_AllDistances_MDS.jpg", width = 3000, height = 1000, res = 90)
alldist.2D.phylumGlom +
  scale_color_manual(values = c('blue', 'red'))+
  theme(axis.text.x = element_text(size = 20),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 30),
        title = element_text(size = 30),
        strip.text = element_text(size = 50))+
  labs(title = "Agglomeration at Phylum level")
dev.off()

jpeg("./Plots/phylumGlom_AllDistances_Datasets.jpg", width = 3000, height = 1000, res = 90)
dataset.2D.phylumGlom +
  theme(axis.text.x = element_text(size = 20),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 30),
        title = element_text(size = 30),
        strip.text = element_text(size = 50))+
  labs(title = "Agglomeration at Phylum level")
dev.off()
```


########
# UMAP #
########

```{r umap-phylumGlom, fig.width=8, fig.height = 4}
library(umap)
library(ggpubr)

umap.phylumGlom <- umap(t(genusTable), n_neighbors=100)
dims_umap.phylumGlom <- umap.phylumGlom$layout
colnames(dims_umap.phylumGlom) <- c("UMAP_1", "UMAP_2")
dims_umap.phylumGlom <- cbind(as.data.frame(dims_umap.phylumGlom), matcol)

a <- ggplot(dims_umap.phylumGlom, aes(x = UMAP_1, y = UMAP_2, color = host_disease))+
  geom_point(size = 1, alpha = 0.7)+
  scale_color_manual(values = c('blue', 'red'))+
  labs(title = 'Agglomeration at Phylum level')

b <- ggplot(dims_umap.phylumGlom, aes(x = UMAP_1, y = UMAP_2, color = author))+
  geom_point(size = 1, alpha = 0.7)

c <- ggplot(dims_umap.phylumGlom, aes(x = UMAP_1, y = UMAP_2, color = sequencing_tech))+
  geom_point(size = 1, alpha = 0.7)+
  scale_color_manual(values = c('#6600FF', '#00CC66', '#FF6633'))

jpeg("./New_Figures/umap_02.jpg", width = 3500, height = 1000, res = 250)
ggarrange(a, b, c, ncol = 3)
dev.off()

```


```{r umap-famGlom, fig.width=8, fig.height = 4}
umap.famGlom <- umap(t(familyTable), n_neighbors=500) #umap
dims_umap.famGlom <- umap.famGlom$layout # extract coordinates
colnames(dims_umap.famGlom) <- c("UMAP_1", "UMAP_2")
dims_umap.famGlom <- cbind(as.data.frame(dims_umap.famGlom), mat_col) # add metadata
dims_umap.famGlom <- dims_umap.famGlom[samples_order,] # reorder samples

a <- ggplot(dims_umap.famGlom, aes(x = UMAP_1, y = UMAP_2, color = host_disease))+
  geom_point(size = 1, alpha = 0.7)+
  scale_color_manual(values = c('blue', 'red'))+
  labs(title = 'Agglomeration at Family level')

b <- ggplot(dims_umap.famGlom, aes(x = UMAP_1, y = UMAP_2, color = author))+
  geom_point(size = 1, alpha = 0.7)

c <- ggplot(dims_umap.famGlom, aes(x = UMAP_1, y = UMAP_2, color = sequencing_tech))+
  geom_point(size = 1, alpha = 0.7)+
  scale_color_manual(values = c('#6600FF', '#00CC66', '#FF6633'))

jpeg("./Plots/umap_fam.jpg", width = 3500, height = 1000, res = 250)
ggarrange(a, b, c, ncol = 3)
dev.off()
```


```{r umap-genusGlom, fig.width=10}
#dim(genusTable)
#dim(ratios)
#table(is.na(ratios))

#test <- scale(ratios, center = TRUE, scale = TRUE)
#boxplot(test)

umap.phyla <- umap(test, n_neighbors=50) #umap on samples (rows) and genera abundances (columns)
dims_umap.phyla<- umap.phyla$layout # extract coordinates
colnames(dims_umap.phyla) <- c("UMAP_1", "UMAP_2")
dims_umap.phyla <- cbind(as.data.frame(dims_umap.phyla), mat_col) # add metadata
dims_umap.phyla <- dims_umap.phyla[samples_order,] # reorder samples
dims_umap.phyla[is.na(dims_umap.phyla$host_disease), "host_disease"] <- "NA"
dims_umap.phyla[dims_umap.phyla$author == "Pozuelo", "sequencing_tech"] <- "Illumina MiSeq single-end"
dims_umap.phyla[dims_umap.phyla$sequencing_tech == "Illumina MiSeq", "sequencing_tech"] <- "Illumina MiSeq paired-end"


a <- ggplot(dims_umap.phyla, aes(x = UMAP_1, y = UMAP_2, color = host_disease))+
  geom_point(size = 5, alpha = 0.7)+
  scale_color_manual(values = c('blue', 'red', 'black'))+
  theme_bw()+
  theme(legend.text = element_blank(),
        legend.title = element_blank(),
        #legend.text = element_text(size = 40),
        #legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(title = 'Agglomeration at Phylum level', color = "Disease")

b <- ggplot(dims_umap.phyla, aes(x = UMAP_1, y = UMAP_2, color = author))+
  geom_point(size = 5, alpha = 0.7)+
  theme_bw()+
  theme(legend.text = element_blank(),
        legend.title = element_blank(),
        #legend.text = element_text(size = 40),
        #legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(color = "Datasets")

c <- ggplot(dims_umap.phyla, aes(x = UMAP_1, y = UMAP_2, color = sequencing_tech))+
  geom_point(size = 5, alpha = 0.7)+
  scale_color_manual(values = c('#6600FF', '#006600', '#66CC00', "#FF6633"))+
  theme_bw()+
  theme(legend.text = element_blank(),
        legend.title = element_blank(),
        #legend.text = element_text(size = 40),
        #legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(color = "Sequencing technology")

jpeg("./Figures/umap_phylum.jpg", width = 1800, height = 4500, res = 200)
ggarrange(a, b, c, nrow = 3)
dev.off()
```

# FAMILY

```{r, fig.width=10}
dim(familyTable)

num1 <- c()
den1 <- c()

for(i in 1:105){
  #cat("num:", num1, "\n")
  num1 <- c(num1, rep(i, times = 106-i))
  #cat("den:", den1, "\n")
  den1 <- c(den1, (i+1):106)
}
length(num1) # 5565
length(den1) # 5565

max(den1)

# Get the numerator, denominator, and rownames for the ratios
num1 <- rownames(familyTable[num1,])
den1 <- rownames(familyTable[den1,])
out1 <- paste0(rownames(familyTable[num1,]), "/", rownames(familyTable[den1,]))

out1[1:5]

length(num1) # 5565
length(den1) # 5565
length(out1) # 5565

# Compute the ratios
ratiosFam <- as.data.frame(familyTable)
ratiosFam[out1,] <- mapply(function(x, y) x/y, familyTable[num1,], familyTable[den1,]) # compute the ratios
ratiosFam <- ratiosFam[-c(1:106),] # remove first 106 rows that are not ratios (it is simply the families)

# Get the log ratios
ratiosFam <- log2(ratiosFam) # Log transform
ratiosFam <- t(ratiosFam) # Get the observations (samples) in rows, and predictors (ratios) in columns
dim(ratiosFam) # 605 rows (samples) and 5565 columns (log2 ratios)
ratiosFam <- ratiosFam[,!colnames(ratiosFam) %in% 'Neisseriaceae/Sedimentibacteraceae']
table(is.na(ratiosFam))

# Mean centering
ratiosFam <- scale(ratiosFam, center = TRUE, scale = TRUE)

umap.fam <- umap(ratiosFam, n_neighbors=100) #umap on samples (rows) and genera abundances (columns)
dims_umap.fam <- umap.fam$layout # extract coordinates
colnames(dims_umap.fam) <- c("UMAP_1", "UMAP_2")
dims_umap.fam <- cbind(as.data.frame(dims_umap.fam), mat_col) # add metadata
dims_umap.fam <- dims_umap.fam[samples_order,] # reorder samples
dims_umap.fam[is.na(dims_umap.fam$host_disease), "host_disease"] <- "NA"
dims_umap.fam[dims_umap.fam$author == "Pozuelo", "sequencing_tech"] <- "Illumina MiSeq single-end"
dims_umap.fam[dims_umap.fam$sequencing_tech == "Illumina MiSeq", "sequencing_tech"] <- "Illumina MiSeq paired-end"
dims_umap.fam$author <- as.character(dims_umap.fam$author)
dims_umap.fam[dims_umap.fam$author == "LoPresti", "author"] <- "Lo Presti"
dims_umap.fam$author <- factor(dims_umap.fam$author, levels = authors_order)

a <- ggplot(dims_umap.fam, aes(x = UMAP_1, y = UMAP_2, color = host_disease))+
  geom_point(size = 5, alpha = 0.7)+
  scale_color_manual(values = c('blue', 'red', 'black'))+
  theme_bw()+
  theme(#legend.text = element_blank(),
        #legend.title = element_blank(),
        legend.text = element_text(size = 40),
        legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(title = 'Agglomeration at Family level', color = "Disease")

b <- ggplot(dims_umap.fam, aes(x = UMAP_1, y = UMAP_2, color = author))+
  geom_point(size = 5, alpha = 0.7)+
  theme_bw()+
  theme(#legend.text = element_blank(),
        #legend.title = element_blank(),
        legend.text = element_text(size = 40),
        legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(color = "Datasets")

c <- ggplot(dims_umap.fam, aes(x = UMAP_1, y = UMAP_2, color = sequencing_tech))+
  geom_point(size = 5, alpha = 0.7)+
  scale_color_manual(values = c('#6600FF', '#006600', '#66CC00', "#FF6633"))+
  theme_bw()+
  theme(#legend.text = element_blank(),
        #legend.title = element_blank(),
        legend.text = element_text(size = 40),
        legend.title = element_text(size = 40),
        panel.border = element_rect(size = 2),
        axis.text = element_text(size = 20),
        axis.title = element_blank())+
  labs(color = "Sequencing technology")

jpeg("./Figures/umap_fam.jpg", width = 1800, height = 4500, res = 200)
ggarrange(a, b, c, nrow = 3)
dev.off()
```

