---
title: "Differential abundance"
output: html_notebook
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE,
                      root.dir = "/Users/enigma/Desktop/Munich/Praktikum/Data/Rproject_COMBINE/PREDICTION/")
```


```{r library}
library(dada2) # manage fastq file
library(ShortRead) # manage fastq file
library(Biostrings) # to manage dna sequences as strings
library(seqTools) # analyse read length in fastq file
library(phyloseq) # for phyloseq object
library(ggplot2) # for plots
library(ggrepel)
library(reshape2)
library(data.table)
library(plyr)
library(dplyr)
library("doParallel")
library("foreach")
library("DECIPHER") # for phylogenetic tree
library("phangorn")
library("plotly") # plot 3D
library("microbiome") # for centered log-ratio
library("coda") # Aitchison distance
library("coda.base") # Aitchison distance
library("vegan") # NMDS
library(pheatmap) # for heat map
library(RColorBrewer)
library(glmnet) # for prediction
library(gtools) # for combinations
library(tidyverse)
library("DESeq2")
```



# IMPORT DATA
```{r import-data}
# Import phyloseq object with all fecal samples with > 1000 total count per sample
physeq <- readRDS("../physeq_best_fecal.rds")
physeq
otu_table(physeq)[1:5,1:5]
sample_data(physeq)

# Import metadata (disease status, author, sequencing technology for each sample)
#metadata <- readRDS('../mat_col.rds')
```

# CONVERT TO DESEQ2

```{r, fig.height = 8, fig.width=10}
# Remove samples for which we don't know the disease status (Ringel dataset)
physeq.best <-  subset_samples(physeq, author != "Ringel-Kulka")
physeq.best # 530 samples
sum(otu_table(physeq.best))


# Add pseudocounts
physeq.pseudocts <- physeq.best
sum(otu_table(physeq.pseudocts))
otu_table(physeq.pseudocts)[otu_table(physeq.pseudocts) == 0] <- 1
physeq.pseudocts

physeq.test <- tax_glom(physeq.best, "Family")
sum(otu_table(physeq.test))
otu_table(physeq.test)[otu_table(physeq.test) == 0] <- 1

deseq <-  phyloseq_to_deseq2(physeq.test, ~ host_disease)
deseq <-  DESeq(deseq, test="Wald", fitType="parametric")

#XXXXXXX
res = results(deseq, cooksCutoff = FALSE)
alpha = 0.01
sigtab = res[which(res$padj < alpha), ]
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(physeq.test)[rownames(sigtab), ], "matrix"))
head(sigtab)
dim(sigtab)

theme_set(theme_bw())
scale_fill_discrete <- function(palname = "Set1", ...) {
    scale_fill_brewer(palette = palname, ...)
}
x <- NULL
# Phylum order
x = tapply(sigtab$log2FoldChange, sigtab$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtab$Phylum = factor(as.character(sigtab$Phylum), levels=names(x))
# Class order
x = tapply(sigtab$log2FoldChange, sigtab$Class, function(x) max(x))
x = sort(x, TRUE)
sigtab$Class = factor(as.character(sigtab$Class), levels=names(x))
# Order order
x = tapply(sigtab$log2FoldChange, sigtab$Order, function(x) max(x))
x = sort(x, TRUE)
sigtab$Order = factor(as.character(sigtab$Order), levels=names(x))
# Family order
x = tapply(sigtab$log2FoldChange, sigtab$Family, function(x) max(x))
x = sort(x, TRUE)
sigtab$Family = factor(as.character(sigtab$Family), levels=names(x))
# Genus order
x = tapply(sigtab$log2FoldChange, sigtab$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtab$Genus = factor(as.character(sigtab$Genus), levels=names(x))

ggplot(sigtab, aes(x=Order, y=log2FoldChange, color=Phylum)) + geom_point(size=6) + 
  theme(axis.text.x = element_text(size = 20, angle = -90, hjust = 0, vjust=0.5))
```

