---
title: "Pozuelo2015_DataAnalysis"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r, include=FALSE}
knitr::opts_chunk$set(fig.width = 15, fig.height = 5, warning=FALSE, message=FALSE,
                      root.dir = "/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/")
```



#######################
## IMPORT LIBRAIRIES ##
#######################
```{r library-import, echo=FALSE}
library(dada2) # manage fastq file

library(ShortRead) # manage fastq file

library(Biostrings) # to manage dna sequences as strings

library(seqTools) # analyse read length in fastq file

library(phyloseq) # for phyloseq object

library(ggplot2) # for plots
library(ggrepel)

library(data.table)

library(plyr)
library(dplyr)

library("doParallel")

library("foreach")

library("DECIPHER") # for phylogenetic tree

library("phangorn")

library("plotly") # plot 3D

library("microbiome") # for centered log-ratio

library("coda") # Aitchison distance

library("coda.base") # Aitchison distance

library("vegan") # NMDS

library(pheatmap) # for heat map
```


#################
## IMPORT DATA ##
#################

XXXX
```{r TO-DELETE, eval = FALSE, echo = TRUE}
# Import metadata
metadata_table <- read.table("/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/PozueloSraRunTable.txt", header = TRUE,
                             sep = ",")
# Keep only relevant metadata
metadata_table <- metadata_table[, c("Run", "Sample.Name")]
rownames(metadata_table) <- metadata_table$Run # assign sample names as row names
metadata_table$Sample.Name <- as.character(metadata_table$Sample.Name) # Sample.Name is of class "factor", transform it to characters
metadata_table$Collection <- "1st" # add column to specify if sample is from first or 2nd collection
metadata_table[gsub('.*(?=.{2}$)', '', metadata_table$Sample.Name, perl=T) == '.2', 'Collection'] <- "2nd"
metadata_table$host_disease <- "Healthy" # add column for host_disease
metadata_table[substr(metadata_table$Sample.Name, 1, 2) == "MO", 'host_disease'] <- "IBS"
metadata_table[, 'Sample.Name'] <- gsub("MO", "IBS", metadata_table[, 'Sample.Name']) # replace 'MO' by 'IBS' in the sample.names
metadata_table$Sample.Name <- gsub("\\.2$", "", metadata_table[, 'Sample.Name']) # remove the '.2' at the end of some of the sample.names (indicating it is the 2nd collection)

colnames(metadata_table) <- c("Run", "individual", "time_point", "host_disease")

metadata_table$author <- "Pozuelo"
metadata_table$sequencing_tech <- "Illumina MiSeq"
metadata_table$variable_region <- "V4"

saveRDS(metadata_table, "/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/metadata_final.rds")


#________________________________________________________________________________________________________
# Import OTU table (rows: sample names // columns: sequence variants)
seqtable.nochim <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/ASVtable_final.rds")
dim(seqtable.nochim) # should have 290 samples and 6222 ASV
rownames(seqtable.nochim) <- gsub("\\.fastq","", rownames(seqtable.nochim)) # remove the .fastq extension in sample names

# Import Taxonomic table (rows: sequence variants // columns: Kingdom, Phylum, Class, ...)
taxa <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/taxa_final.rds")
dim(taxa)
```
XXXXX


```{r import-for-pdf, echo = FALSE}
metadata_table <- readRDS('/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/metadata_final.rds')
physeq <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/OutputPhangorn/physeq_pozuelo.rds")
physeq.rel <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/Rproject_DataAnalysis/physeq_relative.rds")
physeq.clr <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/Rproject_DataAnalysis/physeq_clr.rds")

# Distance methods
dist_methods <- unlist(distanceMethodList)
dist_methods <- dist_methods[-which(dist_methods=="ANY")] # Remove the user-defined distance
dist_methods <- dist_methods[-c(1,15,18)] #remove unweighted unifrac, mountford, chao (not gonna work on pseudocounts)
dist_methods <- dist_methods[1:16] #keep only the 16 first methods (remove the beta diversity distances)
dist_methods <- dist_methods[-2] #dpcoa takes too much time
dist_methods

# For heatmaps: have group color
mat_col <- data.frame(group = metadata_table[,"host_disease"],
                      collection = metadata_table[,"Collection"])
rownames(mat_col) <- rownames(metadata_table)
```



############################
## CREATE PHYLOSEQ OBJECT ##
############################

```{r phyloseq-object, eval = FALSE, echo = TRUE}
#____________________________________________________________________
# OTU TABLE, TAXA TABLE AND METADATA
physeq <- phyloseq(otu_table(seqtable.nochim, taxa_are_rows=FALSE), # by default, in otu_table the sequence variants are in rows
                  sample_data(metadata_table), 
                  tax_table(taxa))

#____________________________________________________________________
# PHYLOGENETIC TREE

# Multiple-sequence alignment (know which regions are conserved/different to be able to do the phylogeny)
seqs <- getSequences(seqtable.nochim) # get the sequence variants (the ASVs)
names(seqs) <- seqs # This propagates to the tip labels of the tree
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA) # by default, anchor = 0.7 which means that 70% of sequences must share a common region to anchor the alignment space.

# Construct a neighbor-joining tree
phang.align <- phyDat(as(alignment, "matrix"), type="DNA") # transform the aligned DNA sequences into a phyDat (phangorn) object
dm <- dist.ml(phang.align) # compute pairwise distance between DNA sequences (with the Jukes-Cantor estimate of the evolutionary distance)
treeNJ <- NJ(dm) # create a neighbor-joining tree estimation based on the distance matrix
fit <- pml(treeNJ, data=phang.align) # get the likelihood of the phylogenetic tree given the sequence alignment (then we'll optimize it)

## negative edges length changed to 0!

# Fit a generalized time-reversible with gamma rate variation (GTR+G+I)
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE, # gamma rate and proportion of variable size get optimized
                      rearrangement = "stochastic", control = pml.control(trace = 0)) # (trace = 0) don't show output during optimization
# stochastic tree rearrangement

# Add tree to physeq
physeq <- merge_phyloseq(physeq, phy_tree(fitGTR$tree))
```

```{r plot-tree, fig.height = 6, fig.width = 10}
# Look at the tree
plot_tree(physeq, color = "host_disease", ladderize="left", sizebase = 2)
```


```{r phyloseq-object-2, eval = FALSE, echo = TRUE}
#____________________________________________________________________
# GIVE SURNAMES TO ASVs

# Give surnames to sequence variants & store the sequence variants in "refseq" in the phyloseq object
dna <- DNAStringSet(taxa_names(physeq)) # get the sequence variants (ASVs)
names(dna) <- taxa_names(physeq) # no idea what this does
physeq <- merge_phyloseq(physeq, dna) # store the dna sequences in the refseq of the phyloseq object
taxa_names(physeq) <- paste0("OTU", seq(ntaxa(physeq))) # replace the whole dna sequences in the taxa_names by a surname OTU1, OTU2, etc.

# Save physeq object
saveRDS(physeq, "/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/Rproject_DataAnalysis/physeq.rds")
saveRDS(physeq, "/Users/scarcy/Desktop/Praktikum/Data/Rproject_COMBINE/physeq_pozuelo.rds")
```



######################################
## ABSOLUTE AND RELATIVE ABUNDANCES ##
######################################
```{r plot-abundances, echo = TRUE, fig.height = 6}
# Relative abundance for Phylum
phylum.table <- physeq %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

# Relative abundance for Class
class.table <- physeq %>%
  tax_glom(taxrank = "Class") %>%                     # agglomerate at class level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format


# Plot Phylum
plot_bar(physeq, fill = "Phylum") + facet_wrap("host_disease", scales="free") +
  theme(axis.text.x = element_text(size = 8))+
  labs(x = "Samples", y = "Total read count", title = "Pozuelo dataset (2015)")+
  ylim(0, 45000)

ggplot(phylum.table[phylum.table[,'Kingdom'] == 'Bacteria',], aes(x = Sample.Name, y = Abundance, fill = Phylum))+
  facet_wrap(~ host_disease + Collection, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 6, angle = -90))+
  labs(x = "Samples", y = "Relative abundance", title = "1st and 2nd time points")

# Plot Class
plot_bar(physeq, fill = "Class")+ facet_wrap("host_disease", scales="free") +
  theme(axis.text.x = element_text(size = 8))+
  labs(x = "Samples", y = "Total read count", title = "Pozuelo dataset (2015)")+
  ylim(0, 45000)

ggplot(class.table[class.table[,'Kingdom'] == 'Bacteria',], aes(x = Sample, y = Abundance, fill = Class))+
  facet_wrap(~ host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 8, angle = -90))+
  labs(x = "Samples", y = "Relative abundance", title = "Pozuelo dataset (2015)")

```


```{r plot-log-ratio, fig.height = 4, fig.width=7}
# Extract abundance only Bacteroidota and Firmicutes
bacter <- phylum.table[phylum.table[,'Phylum'] == 'Bacteroidota', c('Sample', 'Abundance', 'host_disease',
                                                                    'Collection', 'Phylum')]
bacter <- bacter[order(bacter$Sample),] # order by sample name
firmi <- phylum.table[phylum.table[,'Phylum'] == 'Firmicutes', c('Sample', 'Abundance', 'host_disease',
                                                                 'Collection','Phylum')]
firmi <- firmi[order(firmi$Sample),] # order by sample name

# Calculate log2 ratio Firmicutes/Bacteroidota
ratio_FB <- data.table('Sample' = bacter$Sample,
                       'host_disease' = bacter$host_disease,
                       'Collection' = bacter$Collection,
                       'Bacteroidota' = bacter$Abundance,
                       'Firmicutes' = firmi$Abundance)
ratio_FB[,"Log_ratio_FB"] <- log2(ratio_FB[,"Firmicutes"] / ratio_FB[,"Bacteroidota"])

# Plot log2 ratio Firmicutes/Bacteroidota
par(mar = c(3, 10, 3, 10))
boxplot(data = ratio_FB, Log_ratio_FB ~ host_disease, ylab = "Log2 Ratio Abundance", xlab = "", main = "Firmicutes to Bacteroidota ratio")

boxplot(data = ratio_FB, Log_ratio_FB ~ Collection, ylab = "Log2 Ratio Abundance", xlab = "", main = "Firmicutes to Bacteroidota ratio")

```


```{r statistics, eval = FALSE, echo = FALSE}
# Total count per sample (mean, min, max)
min(sample_sums(physeq))
max(sample_sums(physeq))
median(sample_sums(physeq))

# Nb of individuals
table(metadata_table$Sample.Name)

# Look at the nb of Phyla, Classes, etc.
table(taxa[taxa[,1] == 'Eukaryota',2])
table(taxa[taxa[,1] == 'Bacteria',5])
colnames(taxa)

mean(phylum.table[phylum.table[,'Phylum'] == 'Verrucomicrobiota', 'Abundance'])
mean(phylum.table[phylum.table[,'Phylum'] == 'Firmicutes' & phylum.table[,'host_disease'] == 'Healthy', 'Abundance'])

mean(class.table[class.table[,'Class'] == 'Vampirivibrionia', 'Abundance'])
```


```{r export-plots-1, eval = FALSE, echo = FALSE}
jpeg("./Plots/plot_phylum_absAbundance.jpg", width = 4000, height = 2000)
plot_bar(physeq, fill = "Phylum") + facet_wrap("host_disease", scales="free") +
  theme(axis.text.x = element_text(size = 30, angle = -90),
        axis.text.y = element_text(size = 30),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70),
        strip.text = element_text(size = 50))+
  labs(x = "Samples", y = "Total read count", title = "Pozuelo dataset (2015)")+
  ylim(0, 45000)
dev.off()

jpeg("./Plots/plot_phylum_relAbundance.jpg", width = 4000, height = 2000)
ggplot(phylum.table[phylum.table[,'Kingdom'] == 'Bacteria',], aes(x = Sample, y = Abundance, fill = Phylum))+
  facet_wrap(~ host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 10, angle = -90),
        axis.text.y = element_text(size = 30),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70),
        strip.text = element_text(size = 50))+
  labs(x = "Samples", y = "Relative abundance", title = "Pozuelo dataset (2015)")
dev.off()

jpeg("./Plots/plot_timePoints_relAbundance.jpg", width = 4000, height = 2000)
ggplot(phylum.table[phylum.table[,'Kingdom'] == 'Bacteria',], aes(x = Sample.Name, y = Abundance, fill = Phylum))+
  facet_wrap(~ host_disease + Collection, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 15, angle = -90),
        axis.text.y = element_text(size = 30),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70),
        strip.text = element_text(size = 50))+
  labs(x = "Samples", y = "Relative abundance", title = "Pozuelo dataset (2015): 1st and 2nd time points")
dev.off()

jpeg("./Plots/plot_class_relAbundance.jpg", width = 4000, height = 2000)
ggplot(class.table[class.table[,'Kingdom'] == 'Bacteria',], aes(x = Sample, y = Abundance, fill = Class))+
  facet_wrap(~ host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 10, angle = -90),
        axis.text.y = element_text(size = 30),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70),
        strip.text = element_text(size = 50))+
  labs(x = "Samples", y = "Relative abundance", title = "Pozuelo dataset (2015)")
dev.off()

jpeg("./Plots/Log2Ratio_FirmBact.jpg", width = 1500, height = 2000)
ggplot(ratio_FB, aes(x = host_disease, y = Log_ratio_FB))+
  geom_boxplot(outlier.size = 4, lwd = 3)+
  labs(x = "",  y = 'Log2 Firmicutes/Bacteroidota Abundance Ratio', title = "Pozuelo dataset (2015)")+
  theme_classic()+
  theme(axis.text = element_text(size = 50),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70))
dev.off()

```



####################
## NORMALIZE DATA ##
####################

```{r normalize-data, eval = FALSE, echo = TRUE}
# REMOVE SAMPLES WITH LESS THAN 1000 COUNTS
physeq_best <-  prune_samples(sample_sums(physeq)>=1000, physeq)
physeq_best <- subset_taxa(physeq_best, Kingdom == 'Bacteria' | Kingdom == 'Archaea') # remove the 2 NA Kingdoms and Eukaryota
physeq_best <- subset_taxa(physeq_best, !is.na(Phylum)) # remove the 105 NA Phyla
physeq_best

#____________________________________________________________________
# PHYLOSEQ OBJECT WITH NON-ZERO COMPOSITIONS
physeq.NZcomp <- physeq_best
otu_table(physeq.NZcomp)[otu_table(physeq.NZcomp) == 0] <- 0.5

# check the 0 values have been replaced
otu_table(physeq_best)[1:5,1:5]
otu_table(physeq.NZcomp)[1:5,1:5]

# transform into compositions
physeq.NZcomp <- transform_sample_counts(physeq.NZcomp, function(x) x / sum(x) ) # divide each count by the total number of counts (per sample)
sum(otu_table(physeq.NZcomp) < 0) # see how many negative values are present in the matrix
table(rowSums(otu_table(physeq.NZcomp))) # check if there is any row not summing to 1

# save the physeq.NZcomp object
saveRDS(physeq.NZcomp, "/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/Rproject_DataAnalysis/physeq_NZcomp.rds")

#____________________________________________________________________
# PHYLOSEQ OBJECT WITH RELATIVE COUNT (BETWEEN 0 AND 1)
physeq.rel <- physeq_best
physeq.rel <- transform_sample_counts(physeq.rel, function(x) x / sum(x) ) # divide each count by the total number of counts (per sample)

# check the counts are all relative
otu_table(physeq)[1:5, 1:5]
otu_table(physeq.rel)[1:5, 1:5]

# sanity check
sum(otu_table(physeq.rel) < 0) # see how many negative values are present in the matrix
sum(rowSums(otu_table(physeq.rel)) == 1) # check if there is any row not summing to 1

# save the physeq.rel object
saveRDS(physeq.rel, "/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/Rproject_DataAnalysis/physeq_relative.rds")

#____________________________________________________________________
# PHYLOSEQ OBJECT WITH COMMON-SCALE NORMALIZATION
physeq.CSN <- physeq_best
min(sample_sums(physeq.CSN)) # sanity check
physeq.CSN <- transform_sample_counts(physeq.CSN, function(x) (x*min(sample_sums(physeq.CSN))) / sum(x) ) # divide each count by the total number of counts (per sample)

# sanity check
sum(otu_table(physeq.CSN) < 0) # see how many negative values are present in the matrix
table(rowSums(otu_table(physeq.CSN))) # check if there is any row not summing to 1

# save the physeq.CSN object
saveRDS(physeq.CSN, "/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/Rproject_DataAnalysis/physeq_CSN.rds")


#____________________________________________________________________
# PHYLOSEQ OBJECT WITH CENTERED LOG RATIO COUNT
physeq.clr <- physeq_best
physeq.clr <- transform(physeq.clr, "clr") # the function adds pseudocounts itself

# Compare the otu tables in the original phyloseq object and the new one after CLR transformation
otu_table(physeq_best)[1:5, 1:5] # should contain absolute counts
otu_table(physeq.clr)[1:5, 1:5] # should all be relative

# save the physeq.rel object
saveRDS(physeq.clr, "/Users/scarcy/Desktop/Praktikum/Data/Pozuelo_2015/Rproject_DataAnalysis/physeq_clr.rds")
```



#######################
## COMPUTE DISTANCES ##
#######################

```{r dist-thesis, fig.width = 15, fig.height = 4}
#____________________________________________________________________________________
# Measure distances
Distances <- function(){
  set.seed(123) # for unifrac, need to set a seed
  glom.UniF <-  UniFrac(physeq.rel, weighted=TRUE, normalized=TRUE) # weighted unifrac
  glom.ait <- distance(physeq.clr, method = 'euclidean') # aitchison
  glom.bray <- distance(physeq.CSN, method = "bray") # bray-curtis
  glom.can <- distance(physeq.NZcomp, method = "canberra") # canberra
  dist.list <- list("UniF" = glom.UniF, "Ait" = glom.ait, "Canb" = glom.can, "Bray" = glom.bray)
  
  return(dist.list)
}
#subset_samples(physeq.clr, sample_names(physeq.clr) != 'SRR5245605')

DistancesPlot <- function(d){
  plist <- NULL
  plist <- vector("list", 4)
  names(plist) <- c("Weighted Unifrac", "Aitchison", "Bray-Curtis", "Canberra")
  
  print("Unifrac")
  # Weighted UniFrac
  set.seed(123)
  iMDS.UniF <- ordinate(physeq.rel, "MDS", distance=d$UniF)
  plist[[1]] <- plot_ordination(physeq.rel, iMDS.UniF, color="host_disease")
  
  print("Aitchison")
  # Aitchison
  set.seed(123)
  iMDS.Ait <- ordinate(physeq.clr, "MDS", distance=d$Ait)
  plist[[2]] <- plot_ordination(physeq.clr, iMDS.Ait, color="host_disease")
  
  print("Bray")
  # Bray-Curtis
  set.seed(123)
  iMDS.Bray <- ordinate(physeq.CSN, "MDS", distance=d$Bray)
  plist[[3]] <- plot_ordination(physeq.CSN, iMDS.Bray, color="host_disease")
  
  print("Canberra")
  # Canberra
  set.seed(123)
  iMDS.Can <- ordinate(physeq.NZcomp, "MDS", distance=d$Can)
  plist[[4]] <- plot_ordination(physeq.NZcomp, iMDS.Can, color="host_disease")
  
  # Creating a dataframe to plot everything
  plot.df = ldply(plist, function(x) x$data)
  names(plot.df)[1] <- "distance"
  
  return(plot.df)
}

dist.thesis <- Distances()
plot.df <- DistancesPlot(dist.thesis)
plot.df$author <- "Pozuelo"

p1 <- ggplot(plot.df[plot.df$Collection == "1st" & plot.df$distance == "Weighted Unifrac",], aes(Axis.1, Axis.2, color=host_disease))+
  geom_point(size=12, alpha=0.4)  + scale_color_manual(values = c('blue', 'red'))+
  facet_grid(author~distance, scales='free')+
  theme_bw()+
  theme(axis.title = element_blank(),
        #axis.title.x = element_text(size = 40),
        axis.text = element_text(size = 30),
        legend.text = element_text(size = 40),
        legend.title = element_text(size = 40),
        strip.background = element_rect(fill = "white", size = 3),
        panel.border = element_rect(size = 2),
        #strip.text.x = element_text(size = 50, margin = margin(1,0,1,0, "cm")),
        strip.text.x = element_blank(),
        strip.text.y = element_text(size = 70, margin = margin(0,1,0,1, "cm")),
        #axis.line = element_line(size = 1, arrow = arrow(length = unit(0.1, "inches")))
        )+
  labs(color="Disease")

p2 <- ggplot(plot.df[plot.df$Collection == "1st" & plot.df$distance != "Weighted Unifrac",], aes(Axis.1, Axis.2, color=host_disease))+
  geom_point(size=12, alpha=0.4)  + scale_color_manual(values = c('blue', 'red'))+
  facet_wrap(~distance, scales='free', nrow = 1)+
  guides(color = FALSE)+
  theme_bw()+
  theme(axis.text = element_text(size = 30),
        axis.title.y = element_text(size = 40),
        axis.title.x = element_blank(),
        strip.background = element_rect(fill = "white", size = 1),
        panel.border = element_rect(size = 2),
        #strip.text = element_text(size = 50, margin = margin(1,0,1,0, "cm")),
        strip.text.x = element_blank(),
        #axis.line = element_line(size = 1, arrow = arrow(length = unit(0.1, "inches")))
        )

library(ggplot2)
library(ggpubr)

jpeg("/Users/scarcy/Desktop/Praktikum/Thesis/dist_Pozuelo.jpeg", height = 800, width = 4000, res = 100)
ggarrange(p2, p1, ncol = 2, widths = c(2,1))
dev.off()
```


```{r}
# Function that will allow to color the leaves of the tree according to sample type (biopsy vs stool)
color_leaf<<-function(n){
    if(is.leaf(n)){
        # take the current attributes
        a=attributes(n)
        
        # deduce the line in the original data, to get the corresponding sampleType status
        line=match(attributes(n)$label, plot.df$Run)
        disease=plot.df[line, "host_disease"];
            if(disease=="Healthy"){col_disease="blue"};if(disease=="IBS"){col_disease="red"}
        
        #Modification of leaf attribute
        attr(n,"nodePar") <- c(a$nodePar, list(cex=1.5, lab.cex=0.6, pch=30, col=col_disease, lab.font=1, lab.cex=1))
        }
    return(n)
}
test <- distance(subset_samples(physeq.NZcomp, Collection == "1st"), method = "canberra") # canberra

```

```{r, fig.height = 5, fig.width = 3}
hc.ward <- hclust(test, method = "ward.D") #clustering 
dend.ward <- as.dendrogram(hc.ward) # Make into dendograms

dend.ward <- dendrapply(dend.ward, color_leaf) # Apply the color attributes to the dendograms

jpeg("/Users/scarcy/Desktop/Praktikum/Thesis/hclust_pozuelo.jpg", height = 3500, width = 1500, res = 200)

plot(dend.ward , main="Hierarchical clustering on Canberra distance (ward linkage)", horiz = TRUE, leaflab = "none",
     edgePar = list(lwd = 2))
axis(side = 1, lwd = 4, cex = 4)
legend(8, 190, 
     legend = c("Healthy" , "IBS"), 
     col = c("blue", "red"), 
     pch = c(20,20), bty = "n",  pt.cex = 1.5, cex = 1, 
     text.col = "black", horiz = FALSE, inset = c(0, 0.1))
dev.off()
```


```{r dist-functions, eval = TRUE, echo = TRUE}
#____________________________________________________________________________________
# Measure distances
Distances <- function(physeq_obj){
  set.seed(123) # for unifrac, need to set a seed
  glom.UniF <-  UniFrac(physeq_obj, weighted=TRUE, normalized=TRUE) # weighted unifrac
  glom.ait <- dist(x = otu_table(physeq.clr), method = 'euc') # aitchison
  glom.bray <- phyloseq::distance(physeq_obj, method = "bray") # bray-curtis
  glom.can <- phyloseq::distance(physeq_obj, method = "canberra") # canberra
  glom.gower <- phyloseq::distance(physeq_obj, method = "gower") # gower
  glom.bino <- phyloseq::distance(physeq_obj, method = "binomial") # binomial
  
  dist.list <- list("UniF" = glom.UniF, "Ait" = glom.ait, "Canb" = glom.can, "Bray" = glom.bray,
                    "Gower" = glom.gower, "Binomial" = glom.bino)
  
  return(dist.list)
}


#____________________________________________________________________________________
# Plot 2D ordination
MDS_2D <- function(physeq_obj, ait_dist){
  
  plist <- NULL
  plist <- vector("list", length(dist_methods)+1) # save each plot to a list
  names(plist) <- dist_methods # save the name of each method
  names(plist)[16] <- "aitchison" # save the name of aitchison
  
  # Loop through all distance methods
  for(i in dist_methods){
    # Calculate distance matrix
    #print(i)
    set.seed(123) # in case the distance method needs a rooted tree (weighted unifrac)
    iDist <- phyloseq::distance(physeq_obj, method=i)
    # Calculate ordination
    set.seed(123)
    iMDS  <- ordinate(physeq_obj, "MDS", distance=iDist)
    ## Make plot
    # Don't carry over previous plot (if error, p will be blank)
    p <- NULL
    # Create plot, store as temporary variable, p
    p <- plot_ordination(physeq_obj, iMDS, color="host_disease")
    # Add title to each plot
    p <- p + ggtitle(paste("MDS using distance method ", i, sep=""))
    # Save the graphic to the plot list
    plist[[i]] = p
  }
  
  # Add aitchison
  iMDS  <- ordinate(physeq_obj, "MDS", distance=ait_dist)
  p <- NULL
  p <- plot_ordination(physeq_obj, iMDS, color="host_disease")
  p <- p + ggtitle("MDS using distance method Aitchison")
  plist[[17]] = p
  
  # Creating a dataframe to plot everything
  plot.df = ldply(plist, function(x) x$data)
  names(plot.df)[1] <- "distance"
  
  # Plot
  p.alldist <-  ggplot(plot.df, aes(Axis.1, Axis.2, color=host_disease, shape = Collection))+
    geom_point(size=5, alpha=0.5) +
    scale_color_manual(values = c('blue', 'red'))+
    scale_shape_manual(values = c(16,17))+
    facet_wrap(~distance, scales='free')+
    theme(strip.text = element_text(size = 40),
          legend.text = element_text(size = 20),
          axis.text.x = element_text(size = 20),
          axis.text.y = element_text(size = 20))

  return(p.alldist)
}

#____________________________________________________________________________________
# Plot 3D ordination
MDS_3D <- function(d, name_dist){
  
  # Reset parameters
  mds.3D <- NULL
  xyz <- NULL
  fig.3D <- NULL
  
  # Reduce distance matrix to 3 dimensions
  set.seed(123) # to get the same dimensionality reduction at each run
  mds.3D <- metaMDS(d, method="MDS", k=3, trace = 0)
  xyz <- scores(mds.3D, display="sites") # pull out the x y z coordinates
  
  fig.3D <- plot_ly(x=xyz[,1], y=xyz[,2], z=xyz[,3], type="scatter3d", mode="markers",
                    color=sample_data(physeq.rel)$host_disease, colors = c("blue", "red"))%>%
    #add_trace(x=xyz[,1], y=xyz[,2], z=xyz[,3],
    #       type='scatter3d', mode='text', text = rownames(xyz), textfont = list(color = "black")) %>%
    layout(title = paste('MDS in 3D with', name_dist, 'distance', sep = ' '))
  
  return(fig.3D)
}

```


```{r no-glom1, fig.height = 23, fig.width = 30}
# Get the distances
no_glom.dist <- Distances(physeq_obj = physeq.rel)

# Get the 2D ordination plots
no_glom.2D.alldist <- MDS_2D(physeq.rel, no_glom.dist$Ait)
no_glom.2D.alldist + labs(title = 'All samples') + theme(title = element_text(size = 30))
```


```{r save-distances, eval = FALSE, echo = FALSE}
getwd()

write.csv(as.matrix(no_glom.dist$UniF), file = "dist_wunifrac.csv", row.names = TRUE)
write.csv(as.matrix(no_glom.dist$Ait), file = "dist_aitchison.csv", row.names = TRUE)
write.csv(as.matrix(no_glom.dist$Canb), file = "dist_canberra.csv", row.names = TRUE)

```



```{r no-glom2, fig.height = 6, fig.width = 10, eval = FALSE, echo = TRUE}
# Get 3D MDS plots
no_glom.3D.UniF <- MDS_3D(no_glom.dist$UniF, 'weighted Unifrac')
no_glom.3D.Ait <- MDS_3D(no_glom.dist$Ait, 'Aitchison')
no_glom.3D.Can <- MDS_3D(no_glom.dist$Canb, 'Canberra')
no_glom.3D.Gower <- MDS_3D(no_glom.dist$Gower, 'Gower')
no_glom.3D.Binomial <- MDS_3D(no_glom.dist$Binomial, 'Binomial')

no_glom.3D.UniF
no_glom.3D.Ait
no_glom.3D.Can
no_glom.3D.Gower
no_glom.3D.Binomial
```



#############################
## HIERARCHICAL CLUSTERING ##
#############################

```{r hierarchical-clust, fig.width = 10, fig.width = 7}
Heatmaps <- function(dist_list, fontsize){
  # Weighted Unifrac
  heatmp.UniF <- pheatmap(as.matrix(dist_list$UniF), 
                          clustering_distance_rows = dist_list$UniF,
                          clustering_distance_cols = dist_list$UniF,
                          fontsize = fontsize,
                          fontsize_col = fontsize-5,
                          fontsize_row = fontsize-5,
                          annotation_col = mat_col,
                          annotation_row = mat_col,
                          annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                   collection = c('1st' = '#33FF00', '2nd' = '#9933CC')),
                          cluster_rows = T,
                          cluster_cols = T,
                          clustering_method = 'complete', #hierarchical method
                          main = 'Weighted unifrac distance')

  # Aitchison
  heatmp.Ait <- pheatmap(as.matrix(dist_list$Ait), 
                         clustering_distance_rows = dist_list$Ait,
                         clustering_distance_cols = dist_list$Ait,
                         fontsize = fontsize,
                         fontsize_col = fontsize-5,
                         fontsize_row = fontsize-5,
                         # border_color = NA,
                         cluster_rows = T,
                         cluster_cols = T,
                         clustering_method = "complete", #hierarchical method
                         annotation_col = mat_col,
                         annotation_row = mat_col,
                         annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                   collection = c('1st' = '#33FF00', '2nd' = '#9933CC')),
                         main = "Aitchison distance")
  
    # Canberra
  heatmp.Can <- pheatmap(as.matrix(dist_list$Can), 
                         clustering_distance_rows = dist_list$Can,
                         clustering_distance_cols = dist_list$Can,
                         fontsize = fontsize,
                         fontsize_col = fontsize-5,
                         fontsize_row = fontsize-5,
                         # border_color = NA,
                         cluster_rows = T,
                         cluster_cols = T,
                         clustering_method = "complete", #hierarchical method
                         annotation_col = mat_col,
                         annotation_row = mat_col,
                         annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                  collection = c('1st' = '#33FF00', '2nd' = '#9933CC')),
                         main = "Canberra distance")
  
    # Gower
  heatmp.Gower <- pheatmap(as.matrix(dist_list$Gower), 
                         clustering_distance_rows = dist_list$Gower,
                         clustering_distance_cols = dist_list$Gower,
                         fontsize = fontsize,
                         fontsize_col = fontsize-5,
                         fontsize_row = fontsize-5,
                         # border_color = NA,
                         cluster_rows = T,
                         cluster_cols = T,
                         clustering_method = "complete", #hierarchical method
                         annotation_col = mat_col,
                         annotation_row = mat_col,
                         annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                   collection = c('1st' = '#33FF00', '2nd' = '#9933CC')),
                         main = "Gower distance")
  
  # Binomial
  heatmp.Binomial <- pheatmap(as.matrix(dist_list$Binomial), 
                         clustering_distance_rows = dist_list$Binomial,
                         clustering_distance_cols = dist_list$Binomial,
                         fontsize = fontsize,
                         fontsize_col = fontsize-5,
                         fontsize_row = fontsize-5,
                         # border_color = NA,
                         cluster_rows = T,
                         cluster_cols = T,
                         clustering_method = "complete", #hierarchical method
                         annotation_col = mat_col,
                         annotation_row = mat_col,
                         annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                   collection = c('1st' = '#33FF00', '2nd' = '#9933CC')),
                         main = "Binomial distance")
  
  
  return(list("UniF" = heatmp.UniF, "Ait" = heatmp.Ait, "Can" = heatmp.Can, "Gower" = heatmp.Gower, "Bin" = heatmp.Binomial))
}


# Get the heatmaps
no_glom.heatmaps <- Heatmaps(dist_list = no_glom.dist, fontsize = 8)
```



```{r export-plots-2, eval = FALSE, echo = FALSE}
# MDS 2D all distances
jpeg("./Plots/plot_allDistances_MDS.jpg", width = 3000, height = 2000, res=90)
no_glom.2D.alldist +
  theme(axis.text.x = element_text(size = 20),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 50),
        strip.text = element_text(size = 50))+
  labs(title = "Pozuelo dataset (2015)")
dev.off()

# Heatmaps
jpeg("./Plots/plot_heatmap_aitchison.jpg", width = 2800, height = 2000, res = 300)
no_glom.heatmaps$Ait
dev.off()

jpeg("./Plots/plot_heatmap_UniF.jpg", width = 2800, height = 2000, res = 300)
no_glom.heatmaps$UniF
dev.off()

jpeg("./Plots/plot_heatmap_canberra.jpg", width = 2800, height = 2000, res = 300)
no_glom.heatmaps$Can
dev.off()

jpeg("./Plots/plot_heatmap_binomial.jpg", width = 2800, height = 2000, res = 300)
no_glom.heatmaps$Bin
dev.off()

jpeg("./Plots/plot_heatmap_gower.jpg", width = 2800, height = 2000, res = 300)
no_glom.heatmaps$Gower
dev.off()

```


################################
## REPRODUCE PLOTS FROM PAPER ##
################################

```{r reproduce-plots, fig.height = 3, fig.width=4}
#_____________________________________
# FIGURE 3

#Families
family.table <- physeq %>%
  tax_glom(taxrank = "Family") %>%                     # agglomerate at family level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

ggplot(family.table[family.table[,"Family"] == "Erysipelotrichaceae",],
       aes(x = host_disease, y = Abundance))+
  geom_boxplot()+
  geom_jitter(width = 0.1)+
  labs(x = "",  y = 'Relative Abundance', title = 'Erysipelotrichaceae family')+
  theme_classic()+
  ylim(0,0.02)

ggplot(family.table[family.table[,"Family"] == "Ruminococcaceae",],
       aes(x = host_disease, y = Abundance))+
  geom_boxplot()+
  geom_jitter(width = 0.1)+
  labs(x = "",  y = 'Relative Abundance', title = 'Ruminococcaceae family')+
  theme_classic()

```




