# Pozuelo preprocessing

This folder contains data generated by the [00_Metadata-Pozuelo.Rmd](../../../scripts/analysis-individual/Pozuelo-2015/00_Metadata-Pozuelo.Rmd), [01_Dada2-Pozuelo.Rmd](../../../scripts/analysis-individual/Pozuelo-2015/01_Dada2-Pozuelo.Rmd) and [03_EDA-Pozuelo.Rmd](../../../scripts/analysis-individual/Pozuelo-2015/03_EDA-Pozuelo.Rmd) scripts. More specifically:
- `00_Metadata-Pozuelo/` contains the SRA metadata dataframe (downloaded from the SRA) and the final metadata table computed with the 00_Metadata-Pozuelo script;
- `01_Dada2-Pozuelo/` will contain .rds files saved as checkpoints & also for a quicker html output of the 01_Dada2-Pozuelo.Rmd notebook;
- `03_EDA-Pozuelo/` will contain .rds files saved as checkpoints & also for a quicker html output of the 03_EDA-Pozuelo.Rmd notebook, and it will also contain plots generated from that script;
- `filtered1(revcomp)/` will contain the .fastq files generated from primer removal (in reverse complement orientation);
- `filtered1/` will contain the .fastq files generated from primer removal (in original orientation);
- `filtered2/` will contain the .fastq files generated from quality filtering (in original orientation);
- `original(revcomp)` will contain the raw .fastq files (in reverse complement orientation).

For clarification: in the [01_Dada2-Pozuelo.Rmd](../../../scripts/analysis-individual/Pozuelo-2015/01_Dada2-Pozuelo.Rmd) script, this is how the raw .fastq files are processed:
1. Reverse complement the original .fastq files => saved in [original(revcomp)/](./original(revcomp)/)
2. Trim off reverse primer => saved in [filtered1(revcomp)/](./filtered1(revcomp)/)
3. Reverse complement the reads to get back to original orientation => saved in [filtered1/](./filtered1/)
4. Quality filter the reads => saved in [filtered2/](./filtered2/)
