# Liu preprocessing

This folder contains data generated by scripts in the [scripts/analysis-individual/Liu-2020/](../../../scripts/analysis-individual/Liu-2020/) subfolder. More specifically:
- `00_Metadata-Liu/` contains the SRA metadata dataframe (downloaded from the SRA) and the final metadata table computed with the [00_Metadata-Liu.R](../../../scripts/analysis-individual/Liu-2020/00_Metadata-Liu.R) script;
- `01_Dada2-Liu/` contains .rds files saved as checkpoints & also for a quicker html output of the [01_Dada2-Liu.Rmd](../../../scripts/analysis-individual/Liu-2020/01_Dada2-Liu.Rmd) notebook;
- `03_EDA-Liu/` contains .rds files saved as checkpoints & also for a quicker html output of the [03_EDA-Liu.Rmd](../../../scripts/analysis-individual/Liu-2020/03_EDA-Liu.Rmd) notebook, and it also contains plots generated from that script;
- `filtered1(revcomp)/` contains the .fastq files generated from primer removal (in reverse complement orientation);
- `filtered1/` contains the .fastq files generated from primer removal (in original orientation);
- `filtered2/` contains the .fastq files generated from quality filtering (in original orientation);
- `original(revcomp)` contains the raw .fastq files (in reverse complement orientation).

For clarification: in the [01_Dada2-Liu.Rmd](../../../scripts/analysis-individual/Liu-2020/01_Dada2-Liu.Rmd) notebook, this is how the raw `.fastq` files are processed:
1. Reverse complement the original .fastq files => saved in [original(revcomp)/](./original(revcomp)/)
2. Trim off reverse primer => saved in [filtered1(revcomp)/](./filtered1(revcomp)/)
3. Reverse complement the reads to get back to original orientation => saved in [filtered1/](./filtered1/)
4. Quality filter the reads => saved in [filtered2/](./filtered2/)