# Compute phylogenetic tree

This directory is to compute a phylogenetic tree based on ASV sequences in each individual dataset. Having a phylogenetic tree saved in the phyloseq objects of each dataset can allow to compute the UniFrac distance as &beta;-diversity metric.

We follow the [bioconductor workflow for microbiome data analysis](https://f1000research.com/articles/5-1492) (section "Construct phylogenetic tree"). We recommend you copy this whole directory to a computer cluster to run the R script. 


**Necessary elements**:
1. **Input phyloseq objects** in the `input/` subdirectory containing at least ASV and taxonomic tables. These input phyloseq objects are generated by the scripts `01_Dada2-NameDataset.Rmd` in each individual dataset subdirectory of [analysis-individual](../../../../scripts/analysis-individual/)
3. **R script** to compute phylogenetic trees on all datasets present in the `input/` subdirectory => `Rscript_phylotree_cluster.R` is already provided in this directory, but _make sure to change the file paths at the top of the R script!!_ You can also modify this R script to compute the phylogenetic trees only on datasets of interest (change the `for()` loop at bottom of script).
4. **Bash script** to execute the [R script](./Rscript_phylotree_cluster.R) on your computer cluster (recommended to provide at least XXXXX of virtual memory!). Make sure to modify the [bash script](./bash_phylotree_cluster.sh) provided in this directory to fit your cluster.

<br/>

<p align="center">
	<img src="https://matthewkling.github.io/img/images/tree_metrics.gif" height="500"/>
</p>