---
title: "Dada2_Labus2017"
params:
  date: "!r Sys.Date()"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
---


```{r, include=FALSE}
knitr::opts_chunk$set(fig.width = 3.5, fig.height = 3, warning=FALSE, message=FALSE,
                      root.dir = "~/Projects/MetaIBS/scripts/analysis-individual/Labus-2017/")
```




***********
# 1. IMPORT
***********

```{r library-import}
library(dada2)
packageVersion("dada2") # check dada2 version

library(Biostrings)
library(ShortRead)
library(seqTools) # per base sequence content
library(phyloseq)
library(ggplot2)
library(data.table)
library(plyr)
library(dplyr)
library(qckitfastq) # per base sequence content
library(stringr)

# ROOT DIRECTORY (to modify on your computer)
path.root <- "~/Projects/MetaIBS"
path.labus  <- file.path(path.root, "scripts/analysis-individual/Labus-2017")
path.data   <- file.path(path.root, "data/preprocessing/analysis-individual/Labus-2017")
```


```{r html-import, echo = FALSE}
#__________________________________________________________________________
#____________ THIS IS ONLY FOR (quicker) PDF/HTML OUTPUT __________________
#__________________________________________________________________________

raw_stats     <- readRDS(file.path(path.data, "01_Dada2-Labus/raw_stats.rds")) # seq depth
out1          <- readRDS(file.path(path.data, "01_Dada2-Labus/out1.rds")) # out process (primer removal)
out2          <- readRDS(file.path(path.data, "01_Dada2-Labus/out2.rds")) # out process (QC)
filt1_samples <- sort(list.files(file.path(path.data, "filtered1"), pattern=".fastq.gz", full.names = TRUE)) # reads with primers removed
filt2_samples <- sort(list.files(file.path(path.data, "filtered2"), pattern=".fastq.gz", full.names = TRUE)) # quality-filtered reads
err           <- readRDS(file.path(path.data, "01_Dada2-Labus/err.rds")) # error rates
seq_infered   <- readRDS(file.path(path.data, "01_Dada2-Labus/seq_infered.rds")) # inferred ASVs
taxa          <- readRDS(file.path(path.data, "01_Dada2-Labus/taxa_labus.rds")) # taxa
physeq        <- readRDS(file.path(path.root, "data/analysis-individual/CLUSTER/PhyloTree/input/physeq_labus.rds")) # phyloseq object without phylogenetic tree
```




******************
# 2. QUALITY CHECK
******************

First, we import the fastq files containing the raw reads. The samples were downloaded from the ENA database with the accession number PRJNA373876.

```{r quality-check, echo=TRUE, results="hide", message=FALSE}
# Save the path to the directory containing the fastq zipped files
path.fastq <- file.path(path.labus, "download-Labus-samples/")
# list.files(path) # check we are in the right directory

# fastq filenames have format: SAMPLENAME.fastq.gz
# Saves the whole directory path to each file name
FNs <- sort(list.files(file.path(path.fastq, "original"), pattern=".fastq.gz", full.names = TRUE))
show(FNs[1:5])

# Extract sample names, assuming filenames have format: SAMPLENAME.fastq.gz
sample.names <- sapply(strsplit(basename(FNs), "_"), `[`, 1)
show(sample.names[1:5]) # saves only the file name (without the path)

# Look at quality of all files
for (i in 1:4){ # 1:length(FNs)
  show(plotQualityProfile(FNs[i]))
}

# Look at number of reads per sample
# raw_stats <- data.frame('sample' = sample.names,
#                         'reads' = fastqq(FNs)@nReads)
# min(raw_stats$reads)
# max(raw_stats$reads)
# mean(raw_stats$reads)
```

We will have a quick peak at the per base sequence content of the reads in some samples, to make sure there is no anomaly (i.e. all reads having the same sequence).

```{r per-base-seq-content, fig.width = 6, fig.height = 3, message=FALSE}
# Look at per base sequence content
fseq <- seqTools::fastqq(FNs[10])
rc <- read_content(fseq)
plot_read_content(rc) + labs(title = "Per base sequence content")
plot_read_content(rc) + xlim(0,50) + labs(title = "Per base sequence content")
```




**********************
# 3.  LOOK FOR PRIMERS
**********************

Now, we will look whether the reads still contain the primers. In the paper, they describe using V3-V5 primers, as in the article _Evaluation of 16S rDNA-based community profiling for human microbiome research_ (Human Microbiome Project). Supplementary data in this article provides the primers sequences.

```{r primer-check, eval = TRUE, message=FALSE}
# V3-V5
FWD <- "CCTACGGGAGGCAGCAG"  # 357F primer sequence (in reality, it's sequence 5'-341-357- 3')
REV <- "CCGTCAATTCMTTTRAGT" # 926R primer sequence

# Function that, from the primer sequence, will return all combinations possible (complement, reverse complement, ...)
allOrients <- function(primer) {
    require(Biostrings)
    dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
    orients <- c(Forward = dna, Complement = complement(dna), Reverse = reverse(dna), 
        RevComp = reverseComplement(dna))
    return(sapply(orients, toString))  # Convert back to character vector
}

# Get all combinations of the primer sequences
FWD.orients <- allOrients(FWD) # 357F
REV.orients <- allOrients(REV) # 926R
FWD.orients # sanity check
REV.orients

# Function that counts number of reads containing a sequence
primerHits <- function(primer, fn) {
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE, max.mismatch = 2)
    return(sum(nhits > 0))
}

# Look in all samples if we find the primers
for (i in 1:5){
  cat("SAMPLE", sample.names[i], "with total number of", raw_stats[i,'reads'], "reads\n\n")
  # Get a table to know how many times the 357F and 926R primers are found (in how many reads)
  x <-  rbind(ForwardPrimer = sapply(FWD.orients, primerHits, fn = FNs[[i]]), 
              ReversePrimer = sapply(REV.orients, primerHits, fn = FNs[[i]]))
  print(x)
  cat("\n____________________________________________\n\n")
}
```




********************
# 4. FILTER AND TRIM
********************

### 4.1. Primer removal
The reads indeed contain the reverse primer. We will keep only reads containing the primer, and then remove the primer!

```{r primer-filter, eval = FALSE, echo = TRUE}
# KEEP READS WITH PRIMER AND REMOVE PRIMER
# Place filtered files in a filtered/ subdirectory
filt1_samples <- file.path(path.data, "filtered1", paste0(sample.names, "_filt.fastq.gz"))
# Assign names for the filtered fastq.gz files
names(filt1_samples) <- sample.names

# Filter first to take out primer
out1 <- removePrimers(fn = FNs, fout = filt1_samples,
                      primer.fwd = REV.orients['Forward'], # reverse primer is found at the beginning of the reads
                      trim.fwd = TRUE,
                      orient = FALSE, # keep the reads in their original orientation
                      compress = TRUE, verbose = TRUE)

```

```{r primer-filter-check, fig.width = 3, fig.height = 2}
# Primer removal
out1[1:4,]

# Quality profile after primer removal
for (i in 1:4){
  show(plotQualityProfile(filt1_samples[i]))
}
```


### 4.2. Quality filtering
Then, we perform a quality filtering of the reads.

```{r filter-trim, eval = FALSE, echo = TRUE}
# Place filtered files in a filtered/ subdirectory
filt2_samples <- file.path(path.data, "filtered2", paste0(sample.names, "_filt.fastq.gz"))
# Assign names for the filtered fastq.gz files
names(filt2_samples) <- sample.names

# Filter
out2 <- filterAndTrim(fwd = filt1_samples, filt = filt2_samples,
                      maxEE=3, # reads with more than 3 expected errors (sum(10e(-Q/10))) are discarded
                      truncQ=10, # Truncate reads at the first instance of a quality score less than or equal to truncQ.
                      minLen=150, # Discard reads shorter than 150 bp. This is done after trimming and truncation.
                      compress=TRUE,
                      multithread=TRUE,
                      verbose=TRUE)
```

Let's look at the output filtered fastq files as sanity check.

```{r quality-filter-check, fig.width = 3, fig.height = 2}
out2[1:4,] # show how many reads were filtered in each file

# Look at quality profile of filtered files
for (i in 1:4){
  show(plotQualityProfile(filt2_samples[i]))
}
```




**********************
# 5. LEARN ERROR RATES
**********************

Now we will build the parametric error model, to be able to infer amplicon sequence variants (ASVs) later on.

```{r error-rate-estimate, eval = FALSE}
err <- learnErrors(filt2_samples, multithread=TRUE, randomize = TRUE, verbose = 1)
```

The error rates for each possible transition (A→C, A→G, …) are shown. Points are the observed error rates for each consensus quality score. The black line shows the estimated error rates after convergence of the machine-learning algorithm. The red line shows the error rates expected under the nominal definition of the Q-score. Here the estimated error rates (black line) are a good fit to the observed rates (points), and the error rates drop with increased quality as expected.

```{r plot-errors, fig.height=5, fig.width=5}
plotErrors(err, nominalQ = TRUE)
```




*****************************
# 6. CONSTRUCT SEQUENCE TABLE
*****************************

### 6.1. Infer sample composition
The _dada()_ algorithm infers sequence variants based on estimated errors (previous step). Firstly, we de-replicate the reads in each sample, to reduce the computation time.
De-replication is a common step in almost all modern ASV inference (or OTU picking) pipelines, but a unique feature of derepFastq is that it maintains a summary of the quality information for each dereplicated sequence in $quals.

```{r infer-sample-composition, eval = FALSE}
# Prepare empty vector for the infered sequences
seq_infered <- vector("list", length(sample.names))
names(seq_infered) <- sample.names

# Iterate through the 52 samples
for(sampl in sample.names) {
  cat("Processing:", sampl, "\n")
  derep <- derepFastq(filt2_samples[[sampl]]) # dereplicate the reads in the sample
  seq_infered[[sampl]] <- dada(derep, err=err, multithread=TRUE, # default parameters
                               HOMOPOLYMER_GAP_PENALTY=-1, BAND_SIZE=32) # recommended parameters for 454 sequencing
}
```

```{r}
# Inspect the infered sequence variants from sample 1:5
for (i in 1:5){
  print(seq_infered[[i]])
  print("________________")
}
```


### 6.2. Construct ASV table
We can now construct an amplicon sequence variant table (ASV) table, a higher-resolution version of the OTU table produced by traditional methods.

```{r build-seq-table, fig.height=3, fig.width=4}
# Make sequence table from the infered sequence variants
seqtable <- makeSequenceTable(seq_infered)

# We should have 52 samples (52 rows)
dim(seqtable)

# Inspect distribution of ASVs lengths
hist(nchar(getSequences(seqtable)), breaks = 100, xlab = "ASV length", ylab = "Number of ASVs", main="")
```


### 6.3. Remove chimeras
The core dada method corrects substitution and indel errors, but chimeras remain. Fortunately, the accuracy of sequence variants after denoising makes identifying chimeric ASVs simpler than when dealing with fuzzy OTUs. Chimeric sequences are identified if they can be exactly reconstructed by combining a left-segment and a right-segment from two more abundant “parent” sequences.

```{r remove-chimeras}
seqtable.nochim <- removeBimeraDenovo(seqtable, method="consensus", multithread=TRUE, verbose=TRUE)

# Check how many sequence variants we have after removing chimeras
dim(seqtable.nochim)

# Check how many reads we have after removing chimeras (we should keep the vast majority of the reads)
sum(seqtable.nochim)/sum(seqtable)
```




*****************************************
# 7. LOOK AT READS COUNT THROUGH PIPELINE
*****************************************

Sanity check before assigning taxonomy.

```{r reads-filtered-sanity-check}
# Function that counts nb of reads
getN <- function(x) sum(getUniques(x))

# Table that will count number of reads for each process of interest (input reads, filtered reads, denoised reads, non chimera reads)
track <- cbind(out1,
               out2[,2],
               sapply(seq_infered, getN),
               rowSums(seqtable.nochim),
               lapply(rowSums(seqtable.nochim)*100/out1[,1], as.integer))

# Assign column and row names
colnames(track) <- c("input", "primer-filt", "quality-filt", "denoised", "nonchim", "%input->output")
rownames(track) <- sample.names

# Show final table
track
```


```{r seqdepth, echo=FALSE, eval=FALSE, include=FALSE}
# Save nb of reads per sample before / after dada2 pipeline
nb.reads <- data.frame("before" = raw_stats$reads,
                       "after" = rowSums(seqtable.nochim),
                       "dataset" = "Labus")
# better alternative? choose!!
# nb.reads <- as.data.frame(rowSums(seqtable.nochim)) %>%
#   rename("after"="rowSums(seqtable.nochim)") %>%
#   mutate("Run"=names(rowSums(seqtable.nochim))) %>%
#   right_join(raw_stats %>% rename(Run=sample, before=reads),
#              by="Run") %>%
#   relocate(before) %>%
#   relocate(Run) %>%
#   mutate(after = nafill(after, fill=0),
#          "dataset" = "AGP")

saveRDS(nb.reads, file.path(path.root, "data/preprocessing/dada2-filt/nbreads_labus.rds"))
```




********************
# 8. ASSIGN TAXONOMY
********************

Extensions: The dada2 package also implements a method to make species level assignments based on exact matching between ASVs and sequenced reference strains. Recent analysis suggests that exact matching (or 100% identity) is the only appropriate way to assign species to 16S gene fragments. Currently, species-assignment training fastas are available for the Silva and RDP 16S databases. To follow the optional species addition step, download the silva_species_assignment_v138.fa.gz file, and place it in the directory with the fastq files.

```{r assign-taxonomy-on-cluster, eval=FALSE, echo=FALSE, include=FALSE}
# In case you prefer to assign taxonomy on a computer cluster (in other words run this chunk on a computer cluster), you can export the ASV table
# saveRDS(seqtable.nochim, file.path(path.root, "data/preprocessing/analysis-individual/CLUSTER/taxonomy/input/seqtablenochim_labus.rds"))

# then you can :
# 1. copy the "data/preprocessing/analysis-individual/CLUSTER/taxonomy/" directory to your cluster (where there is the input data & R scripts to assign taxonomy)
# 2. run the R scripts on the cluster (they will output the taxonomic tables in the "output/" subdirectory)
# 3. copy the "taxonomy/output/" subdirectory back to your personal computer
# 4. import the taxonomic table below (in that case you don't need to run the next code chunk "assign-taxonomy"):
# taxa <- readRDS(file.path(path.root, "data/preprocessing/analysis-individual/CLUSTER/taxonomy/output/taxa_labus.rds"))
```

```{r assign-taxonomy, eval = FALSE}
# Assign taxonomy (with silva v138)
taxa <- assignTaxonomy(seqtable.nochim, file.path(path.root, "data/preprocessing/analysis-individual/CLUSTER/taxonomy/silva-taxonomic-ref/silva_nr_v138_train_set.fa.gz"),
                       tryRC = TRUE, # try reverse complement of the sequences
                       multithread=TRUE, verbose = TRUE)

# Add species assignment
taxa <- addSpecies(taxa, file.path(path.root, "data/preprocessing/analysis-individual/CLUSTER/taxonomy/silva-taxonomic-ref/silva_species_assignment_v138.fa.gz"))
```

```{r taxa-sanity-check}
# Check how the taxonomy table looks like
taxa.print <- taxa
rownames(taxa.print) <- NULL # Removing sequence rownames for display only
head(taxa.print)
table(taxa.print[,1]) # Show the different kingdoms (should be only bacteria)
table(taxa.print[,2]) # Show the different phyla
table(is.na(taxa.print[,2])) # is there any NA phyla?
```




***********************
# 9. LAST PREPROCESSING
***********************

We will remove any sample with less than 500 reads from further analysis, and also any ASVs with unassigned phyla.

### 9.1. Create phyloseq object
The preprocessing will be easier to do with ASV, taxonomic and metadata tables combined in a phyloseq object.

```{r phyloseq-preprocess, echo=TRUE, eval=FALSE}
#_________________________
# Import metadata
metadata_table <- read.csv("~/Projects/MetaIBS/analysis-individual/Labus-2017/Metadata-Labus.csv")

# Keep relevant covariates
sampledf <- metadata_table
sampledf$X <- NULL
rownames(sampledf) <- sampledf$Run # assign sample names as row names
sampledf[sampledf$host_disease == "None", "host_disease"] <- "Healthy"

sampledf$sample_type <- "stool"
sampledf$Collection <- "1st"

# Remove .fastq.gz extension in seqtable.nochim
rownames(seqtable.nochim) <- gsub(".fastq.gz", "", rownames(seqtable.nochim))

#_________________________
# Create phyloseq object
physeq <- phyloseq(otu_table(seqtable.nochim, taxa_are_rows=FALSE), # by default, in otu_table the sequence variants are in rows
                  sample_data(sampledf), 
                  tax_table(taxa))

# Remove taxa that are eukaryota, or have unassigned Phyla
physeq <- subset_taxa(physeq, Kingdom != "Eukaryota")
physeq <- subset_taxa(physeq, !is.na(Phylum))
# Remove samples with less than 500 reads
physeq <- prune_samples(sample_sums(physeq)>=500, physeq)
# Some taxa might have been present only in these low-count samples, so we will make sure to remove taxa that are absent in all samples
physeq <- prune_taxa(taxa_sums(physeq)>0, physeq)
```


### 9.2. Quick peek at data analysis
```{r plot, fig.width = 10, fig.height = 4}
# Absolute abundance
plot_bar(physeq, fill = "Phylum")+ facet_wrap("host_disease", scales="free_x") + theme(axis.text.x = element_blank())

# Relative abundance for Phylum
phylum.table <- physeq %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

ggplot(phylum.table, aes(x = Sample, y = Abundance, fill = Phylum))+
  facet_wrap(~ host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 5, angle = -90))+
  labs(x = "Samples", y = "Relative abundance")
```


### 9.3. Save to disk
```{r save, eval=FALSE}
# Save to disk
saveRDS(raw_stats,   file.path(path.data, "01_Dada2-Labus/raw_stats.rds"))
saveRDS(out1,        file.path(path.data, "01_Dada2-Labus/out1.rds"))
saveRDS(out2,        file.path(path.data, "01_Dada2-Labus/out2.rds"))
saveRDS(err,         file.path(path.data, "01_Dada2-Labus/err.rds"))
saveRDS(seq_infered, file.path(path.data, "01_Dada2-Labus/seq_infered.rds"))

# Taxa & Phyloseq object
saveRDS(taxa,   file.path(path.data, "01_Dada2-Labus/taxa_labus.rds"))
saveRDS(physeq, file.path(path.root, "data/analysis-individual/CLUSTER/PhyloTree/input/physeq_labus.rds"))
```




******************
# 10. SESSION INFO
******************

```{r session-info}
sessionInfo()
```