---
title: "UMAP"
output: html_notebook
---


##########
# IMPORT #
##########

## Import libraries

```{r import-libraries, echo = FALSE}
library(phyloseq)
library(ggplot2)
library(plotly)
library(pals)
library(umap)
library(tidyverse)
library(reshape2)
library(gtools)
```


## Import data

```{r import-data}
path.phy <- "~/Projects/IBS_Meta-analysis_16S/data/analysis-individual/CLUSTER/PhyloTree/input"
physeq.labus    <- readRDS(file.path(path.phy, "physeq_labus.rds"))
physeq.lopresti <- readRDS(file.path(path.phy, "physeq_lopresti.rds"))
physeq.ringel   <- readRDS(file.path(path.phy, "physeq_ringel.rds"))
physeq.agp      <- readRDS(file.path(path.phy, "physeq_agp.rds"))
physeq.liu      <- readRDS(file.path(path.phy, "physeq_liu.rds"))
physeq.pozuelo  <- readRDS(file.path(path.phy, "physeq_pozuelo.rds"))
physeq.fukui    <- readRDS(file.path(path.phy, "physeq_fukui.rds"))
physeq.hugerth  <- readRDS(file.path(path.phy, "physeq_hugerth.rds"))
physeq.mars     <- readRDS(file.path(path.phy, "physeq_mars.rds"))
physeq.zhu      <- readRDS(file.path(path.phy, "physeq_zhu.rds"))
physeq.zhuang   <- readRDS(file.path(path.phy, "physeq_zhuang.rds"))
physeq.nagel    <- readRDS(file.path(path.phy, "physeq_nagel.rds"))
physeq.zeber    <- readRDS(file.path(path.phy, "physeq_zeber.rds"))
```


## Merge datasets

```{r merge-datasets}
# Merge phyloseq objects
physeq.all <- merge_phyloseq(physeq.ringel,
                             physeq.labus,
                             physeq.lopresti,
                             physeq.pozuelo,
                             physeq.zhuang,
                             physeq.zhu,
                             physeq.hugerth,
                             physeq.fukui,
                             physeq.mars,
                             physeq.liu,
                             physeq.agp,
                             physeq.nagel,
                             physeq.zeber)

# check on sample_data
# colnames(sample_data(physeq.all))
# table(sample_data(physeq.all)$author, useNA="ifany") #ok
# table(sample_data(physeq.all)$sequencing_tech, useNA="ifany") #ok
# table(sample_data(physeq.all)$variable_region, useNA="ifany") # ok
# 
# table(sample_data(physeq.all)$host_disease, useNA="ifany") # ok
# table(sample_data(physeq.all)$host_subtype, useNA="ifany") # ok
# table(sample_data(physeq.all)$sample_type, useNA="ifany") # ok
# 
# table(sample_data(physeq.all)$host_sex, useNA="ifany") # ok
# table(sample_data(physeq.all)$host_age, useNA="ifany") # ok
# table(sample_data(physeq.all)$host_bmi, useNA="ifany") # ok
# table(sample_data(physeq.all)$Collection, useNA="ifany") # ok

# Keep only fecal samples
physeq.fecal <- subset_samples(physeq.all, sample_type == 'stool') # 2,220 samples
physeq.sigmoid <- subset_samples(physeq.all, sample_type == 'sigmoid') # 431 samples

physeq.fecal <- prune_taxa(taxa_sums(physeq.fecal)>0, physeq.fecal) # remove ASVs that are not present anymore
physeq.sigmoid <- prune_taxa(taxa_sums(physeq.sigmoid)>0, physeq.sigmoid)
```




#########################
# UMAP ON CLR TRANSFORM #
#########################

## RUN UMAP AT FAMILY LEVEL
```{r umap-clr-family}
# Agglomerate data to the family level
physeq_fecal.fam <- tax_glom(physeq.fecal, taxrank = "Family")

# CLR-transform data
physeq_fecal.clr <- microbiome::transform(physeq_fecal.fam, "clr") # the function adds pseudocounts itself
# otu_table(physeq_fecal.clr)[1:5,1:5]

# Get the family table: samples as rows x families as columns
familyTable_fecal.clr <- acast(physeq_fecal.clr %>% psmelt(),
                               Sample ~ Family, fun.aggregate=sum, value.var = 'Abundance')
# familyTable_fecal.clr[1:5,1:5]


# RUN UMAP
set.seed(123)
umapFam_fecal.clr <- uwot::umap(familyTable_fecal.clr, # umap on samples (rows) and families abundances (columns)
                                n_neighbors=20, n_components=3, scale=F, n_threads=8)

# Get the (x,y) coordinates from the UMAP
dims.umapFam_fecal_clr <- umapFam_fecal.clr %>% as.data.frame()
colnames(dims.umapFam_fecal_clr) <- c("UMAP_1", "UMAP_2", "UMAP_3")
# table(rownames(dims.umapFam_fecal_clr) == rownames(familyTable_fecal.clr)) # sanity check
```


## PLOT UMAP AT FAMILY LEVEL

```{r umap-clr-family-plot}
# ADD COVARIATES
# Get a covariates table
covariates_fecal <- data.frame(disease = sample_data(physeq_fecal.clr)[,'host_disease'],
                               subtype = sample_data(physeq_fecal.clr)[,'host_subtype'],
                               seq_tech = sample_data(physeq_fecal.clr)[,'sequencing_tech'],
                               author = sample_data(physeq_fecal.clr)[,'author'],
                               variable_region = sample_data(physeq_fecal.clr)[,'variable_region'],
                               age = sample_data(physeq_fecal.clr)[,'host_age'],
                               bmi = sample_data(physeq_fecal.clr)[,'host_bmi'])
covariates_fecal <- covariates_fecal %>%
  mutate(author = factor(covariates_fecal$author, levels = c('Labus', 'LoPresti', 'Ringel', # 454 pyrosequencing
                                                       'AGP', 'Liu', 'Pozuelo', # Illumina single end
                                                       'Fukui', 'Hugerth', 'Mars', 'Zhu', 'Zhuang', # Illumina paired end
                                                       'Nagel', 'Zeber-Lubecka'))) %>%
  mutate(sequencing_tech = factor(covariates_fecal$sequencing_tech, levels= c("454 pyrosequencing", "Illumina single-end",
                                                                        "Illumina paired-end", "Ion Torrent")))

# Merge covariates to the umap df
# setdiff(rownames(dims.umapFam_fecal_clr), rownames(covariates_fecal)) # sanity check
dims.umapFam_fecal_clr <- merge(as.data.frame(dims.umapFam_fecal_clr), covariates_fecal, by="row.names") # umap+covariates
colnames(dims.umapFam_fecal_clr)[1] <- "Sample"
cat("Nb of samples", nrow(dims.umapFam_fecal_clr), "\n")


# PLOT
# PLOT host_disease
ggplot(dims.umapFam_fecal_clr, aes(x = UMAP_1, y = UMAP_2, color = host_disease))+
  geom_point(size = 2, alpha = 0.7)+
  scale_color_manual(values = c('blue', 'red', 'black'))+
  labs(title = 'Agglomeration at Family level')+
  theme_bw()

# PLOT host_subtype
ggplot(dims.umapFam_fecal_clr, aes(x = UMAP_1, y = UMAP_2, color = host_subtype))+
  geom_point(size = 1, alpha = 0.7)+
  scale_color_manual(values = c('#99CCFF', '#FF3300', '#990000', '#FF66CC','#FFFF66', '#CCCCCC'))+
  labs(title = 'Agglomeration at Family level')+
  theme_bw()

# PLOT author
ggplot(dims.umapFam_fecal_clr, aes(x = UMAP_1, y = UMAP_2, color = author))+
  geom_point(size = 2, alpha = 0.7)+
  scale_color_manual(values=pals::brewer.paired(12), name="")+ # author
  theme_cowplot()

# PLOT seqtech
ggplot(dims.umapFam_fecal_clr, aes(x = UMAP_1, y = UMAP_2, color = sequencing_tech))+
  geom_point(size = 2, alpha = 0.7)+
  scale_color_manual(values = c('#6600FF', '#33CC33', '#006600', '#FF6633'))+
  theme_bw()
plotly::plot_ly(dims.umapFam_fecal_clr, x=~UMAP_1, y=~UMAP_2, z=~UMAP_3, color=~host_subtype, type="scatter3d", mode="markers")
```






###################
# SUPERVISED UMAP #
###################

## GET RATIOS BETWEEN PHYLA
```{r functions-umap}
#________________________________________________________
# Function to agglomerate to given taxonomic level
aggTable <- function(physeq, tax_rank){
  
  cat("\n++ AGGLOMERATE TO", tax_rank, "LEVEL++\n")
  long.agg <- physeq %>%
    tax_glom(taxrank = tax_rank) %>%
    psmelt()
  
  # Get a matrix TaxRank (rows) x Samples (columns)
  if(tax_rank=="Phylum"){agglomeratedTable <- acast(long.agg, Phylum ~ Sample, fun.aggregate=sum, value.var = 'Abundance')}
  else if(tax_rank=="Class"){agglomeratedTable <- acast(long.agg, Class ~ Sample, fun.aggregate=sum, value.var = 'Abundance')}
  else if(tax_rank=="Order"){agglomeratedTable <- acast(long.agg, Order ~ Sample, fun.aggregate=sum, value.var = 'Abundance')}
  else if(tax_rank=="Family"){agglomeratedTable <- acast(long.agg, Family ~ Sample, fun.aggregate=sum, value.var = 'Abundance')}
  else if(tax_rank=="Genus"){agglomeratedTable <- acast(long.agg, Genus ~ Sample, fun.aggregate=sum, value.var = 'Abundance')}
  cat("-> Dimensions matrix:", dim(agglomeratedTable), "\n")
  cat("-> Any 0 counts?", table(agglomeratedTable == 0), "\n")
  
  # Sanity check reshape df
  randomTaxa <- rownames(agglomeratedTable)[10]
  randomSample <- colnames(agglomeratedTable)[785]
  valueAgg <- sum(long.agg[long.agg[,tax_rank] == randomTaxa & long.agg$Sample == randomSample, "Abundance"])
  wideAgg <- agglomeratedTable[randomTaxa, randomSample]
  cat("-> Sanity check: value in long shaped df is", valueAgg, "while in wide shaped df is", wideAgg)
  
  # Save
  # filepath <- paste0("~/IBS/UMAP/data_logratio/", paste0(tolower(tax_rank), "_agg.rds", sep=""), sep="")
  # saveRDS(object=long.agg, file=filepath)
  
  return(agglomeratedTable)
}


#________________________________________________________
# Function to get the table to feed into the UMAP
getTable <- function(physeq, tax_rank){
  
  # Get the agglomerated taxa
  agglomeratedTable <- aggTable(physeq, tax_rank)
  
  # Get log-ratios between taxa
  ratiosTable <- LogRatios(agglomeratedTable, tax_rank)
  
  # Mean-center
  # cat("\n++MEAN-CENTERING...++|n")
  # ratios.scaled <- scale(ratiosTable, center = TRUE, scale = FALSE)
  
  # return(ratios.scaled)
  return(ratiosTable)
}


# Get the data table
ratios.phyla <- getTable(physeq=physeq_fecal.pseudocts, tax_rank = "Phylum")

```



## RUN UMAP

```{r run-supervised-umap}
#________________________________________________________
# Function to run the UMAP
runUMAP <- function(physeq, tax_rank){
  
  # # Get the data table
  # ratios <- getTable(physeq, tax_rank)
  
  # Add host_disease to the ratios matrix
  ratios <- merge(ratios.phyla,
                  data.frame(sequencing_tech = sample_data(physeq_fecal.pseudocts)[,'sequencing_tech']),
                  by="row.names")
  ratios$sequencing_tech <- as.factor(ratios$sequencing_tech)
  rownames(ratios) <- ratios[,"Row.names"]
  ratios[,"Row.names"] <- NULL
  
  cat("\n++RUN UMAP...++\n")
  # Run UMAP
  set.seed(123)
  umap <- uwot::umap(ratios, # umap on samples (rows) and taxa ratios (columns)
                     n_neighbors=50, n_components=3, n_threads=16, y=ratios$sequencing_tech)
  
  # Save
  # filepath <- paste0("~/IBS/UMAP/data_logratio/umap", paste0(tax_rank, ".rds", sep=""), sep="")
  # saveRDS(object=umap, file=filepath)
  
  # Get the (x,y) coordinates from the UMAP
  dims.umap <- umap %>% as.data.frame()
  colnames(dims.umap) <- c("UMAP_1", "UMAP_2", "UMAP_3")
  rownames(dims.umap) <- rownames(ratios)
  
  # Add covariates
  covariates <- data.frame(disease = sample_data(physeq)[,'host_disease'],
                           subtype = sample_data(physeq)[,'host_subtype'],
                           seq_tech = sample_data(physeq)[,'sequencing_tech'],
                           author = sample_data(physeq)[,'author'],
                           variable_region = sample_data(physeq)[,'variable_region'],
                           age = sample_data(physeq)[,'host_age'],
                           bmi = sample_data(physeq)[,'host_bmi'])
  dims.umap <- merge(as.data.frame(dims.umap), covariates, by="row.names") # umap+covariates
  
  # Save
  # filepath <- paste0("~/IBS/UMAP/data_logratio/dims_umap", paste0(tax_rank, ".rds", sep=""), sep="")
  # saveRDS(object=dims.umap, file=filepath)
  
  return(dims.umap)
}

dims.umap <- runUMAP(physeq=physeq_fecal.pseudocts, tax_rank = "Phylum")

```


## PLOT

```{r plot-supervised-umap, fig.width=5, fig.height=3.5}
#### USE PLOTTING FUNCTIONS FROM 02_UMAP_plot.Rmd ####
plot.covariates(plotDF = dims.umap)

# Plot3D
# plot3D(plotDF = dims.umap, coloring="host_disease", colorPal=c("blue", "red", "black"), taxLevel="Phylum")
# plot3D(plotDF = dims.umap, coloring="host_subtype", colorPal=c("#EFF3FF", "#D95F02", "#7570B3", "#E7298A", "#FEEDDE", "grey"), taxLevel="Phylum")
# plot3D(plotDF = dims.umap, coloring="sequencing_tech", colorPal=c('#6600FF', '#33CC33', '#006600', '#FF6633'), taxLevel="Phylum")
# plot3D(plotDF = dims.umap, coloring="author", colorPal=brewer.paired(n=13), taxLevel="Phylum")

# Plot by log ratios or relative abundances
colnames(dims.umap)[1] <- "samples"
umap.df <- merge(dims.umap, ratio.df, by=c("samples", "host_disease", "host_subtype", "sequencing_tech", "author"))
umap.df <- merge(umap.df, relAbundance, by="samples")
  
# Plot log ratios (in 2D)
plot.logRatio()

# Plot 3D
# plot3D(coloring="LogRatio_FirmBact", colorPal=rev(brewer.rdbu(50)), taxLevel="Phylum")
# plot3D(coloring="LogRatio_FirmAct", colorPal=rev(brewer.rdbu(50)), taxLevel="Phylum")
# plot3D(coloring="LogRatio_BactAct", colorPal=rev(brewer.rdbu(50)), taxLevel="Phylum")


# By Phyla relative abundances (in 2D)
plot.relAbundance()

# By Phyla relative abundances (in 3D)
# plot3D(coloring="Firmicutes", colorPal=rev(brewer.rdbu(50)), taxLevel="Phylum")
# plot3D(coloring="Bacteroidota", colorPal=rev(brewer.rdbu(50)), taxLevel="Phylum")
# plot3D(coloring="Actinobacteriota", colorPal=rev(brewer.rdbu(50)), taxLevel="Phylum")
# plot3D(coloring="Proteobacteria", colorPal=rev(brewer.rdbu(50)), taxLevel="Phylum")


# Ruminococcaceae?
familyTable.rel <- physeq.fecal %>%
  tax_glom(taxrank = "Family") %>%
  transform_sample_counts(function(x) {x/sum(x)} ) %>%
  psmelt()

# Extract relative abundance of major phyla
rumino.rel <- familyTable.rel %>%
  filter(Family == "Ruminococcaceae") %>%
  select(Sample, Abundance, Family) %>%
  rename(Ruminococcaceae = Abundance) %>%
  rename(samples = Sample) %>%
  select(-Family)
umap.df <- merge(umap.df, rumino.rel, by="samples")

ggplot(umap.df, aes(x = UMAP_1, y = UMAP_2, color = Ruminococcaceae))+
    geom_point(size=1) +
    scale_color_gradient2(midpoint=mean(umap.df$Ruminococcaceae), low="blue", mid="white", high="red")+
    theme(panel.grid = element_blank(),
          panel.background=element_blank(),
          legend.key.size = unit(0.25, 'cm'),
          legend.text = element_text(size=6),
          legend.title = element_text(size=10),
          axis.title = element_text(size=5),
          axis.text = element_text(size=5))
```





###########################
# REPRODUCE MASTER THESIS #
###########################

```{r master-thesis}
# Get phyloseq object
physeq.small <- merge_phyloseq(physeq.ringel,
                               physeq.labus,
                               physeq.lopresti,
                               physeq.pozuelo,
                               physeq.zhuang,
                               physeq.zhu,
                               physeq.zeber)
# Keep only fecal samples
physeq.small <- subset_samples(physeq.small, sample_type == 'stool') # 612 samples

# Get the familyTable
physeq.small.pseudocts <- physeq.small
otu_table(physeq.small.pseudocts)[otu_table(physeq.small.pseudocts) == 0] <- 0.5
agg.small <- physeq.small.pseudocts %>%
  tax_glom(taxrank = "Family") %>%                     # agglomerate at genus level
  psmelt()                                             # Melt to long format
familyTable.small <- acast(agg.small, Family ~ Sample, value.var = 'Abundance')
# familyTable.small[familyTable.small == 0] <- 10e-7 # add pseudocounts for heatmap
dim(familyTable.small)

# Get log-ratios for family table
ratiosFam.small <- LogRatios(familyTable.small)

# Mean centering
ratiosFam.small <- scale(ratiosFam.small, center = TRUE, scale = TRUE)
table(is.na(ratiosFam.small))

# UMAP
umap.small <- umap(ratiosFam.small, n_neighbors=100) #umap on samples (rows) and families abundances (columns)
dims_umap.small <- umap.small$layout # extract coordinates
colnames(dims_umap.small) <- c("UMAP_1", "UMAP_2")

# Add covariates
mat_col <- data.frame(disease = sample_data(physeq.small)[,'host_disease'],
                        seq_tech = sample_data(physeq.small)[,'sequencing_tech'],
                        author = sample_data(physeq.small)[,'author'])
mat_col[is.na(mat_col$host_disease), "host_disease"] <- "NA"
mat_col[mat_col$author == "Pozuelo", "sequencing_tech"] <- "Illumina MiSeq single-end"
mat_col[mat_col$author %in% c("Zhuang", "Zhu"), "sequencing_tech"] <- "Illumina MiSeq paired-end"
dims_umap.small <- merge(as.data.frame(dims_umap.small), mat_col, by="row.names") # umap+covariates

# PLOT
ggplot(dims_umap.small, aes(x = UMAP_1, y = UMAP_2, color = host_disease))+
  geom_point(size = 1, alpha = 0.7)+
  scale_color_manual(values = c('blue', 'red', 'black'))+
  labs(title = 'Agglomeration at Family level')+
  theme_bw()

ggplot(dims_umap.small, aes(x = UMAP_1, y = UMAP_2, color = author))+
  geom_point(size = 2, alpha = 0.7)+
  theme_bw()

ggplot(dims_umap.small, aes(x = UMAP_1, y = UMAP_2, color = sequencing_tech))+
  geom_point(size = 1, alpha = 0.7)+
  scale_color_manual(values = c('#6600FF', '#33CC33', '#006600', '#FF6633'))+
  theme_bw()
```

