---
title: "Labus2017_DataAnalysis"
output:
  html_document:
    df_print: paged
    toc: true
  pdf_document: default
---

```{r, include=FALSE}
knitr::opts_chunk$set(fig.width = 15, fig.height = 5, warning=FALSE, message=FALSE,
                      root.dir = "~/Projects/IBS_Meta-analysis_16S/analysis-individual/Labus-2017/")
```



############
## IMPORT ##
############

## Librairies

```{r library-import}
library(phyloseq) # for phyloseq object
library(ggplot2)
library(cowplot)
library(plyr)
library(dplyr)
library("plotly") # plot 3D
library("microbiome") # for centered log-ratio
library("coda") # Aitchison distance
library("coda.base") # Aitchison distance
library("vegan") # NMDS
library(pheatmap) # for heatmap


# library(Biostrings) # to manage dna sequences as strings
# library(seqTools) # analyse read length in fastq file
# library(phyloseq) # for phyloseq object
# library(ggplot2) # for plots
# library(ggrepel)
# library(data.table)
# library(plyr)
# library(dplyr)
# 
# library("doParallel")
# library("foreach")
# library("DECIPHER") # for phylogenetic tree
# library("phangorn")
# 
# library("plotly") # plot 3D
# 
# library("microbiome") # for centered log-ratio
# library("coda") # Aitchison distance
# library("coda.base") # Aitchison distance
# library("vegan") # NMDS
# library(pheatmap) # for heat map
```

## Data

```{r import-data}
# Set path
path <- "~/Projects/IBS_Meta-analysis_16S"

# Import phyloseq object
physeq.labus <- readRDS(file.path(path, "phyloseq-objects/physeq_labus.rds"))

# Sanity check
physeq.labus
```


Phylogenetic tree was computed with the package `phangorn`, and the script was run on a cluster. Let's check we have correctly generated a phylogenetic tree.

```{r plot-tree, fig.height = 6, fig.width = 10}
# Look at the tree
plot_tree(physeq.labus, color = "host_disease", ladderize="left")
```


```{r import-for-html, echo = FALSE}
#__________________________________________________________________________
#____________ THIS IS ONLY FOR (quicker) PDF/HTML OUTPUT __________________
#__________________________________________________________________________

# Import phyloseq objects for computing distances
physeq.rel <- readRDS(file.path(path, "data/analysis-individual/Zhuang-2018/02_EDA-Zhuang/physeq_relative.rds"))
physeq.clr <- readRDS(file.path(path, "data/analysis-individual/Zhuang-2018/02_EDA-Zhuang/physeq_clr.rds"))
physeq.CSN <- readRDS(file.path(path, "data/analysis-individual/Zhuang-2018/02_EDA-Zhuang/physeq_CSN.rds"))
physeq.NZcomp <- readRDS(file.path(path, "data/analysis-individual/Zhuang-2018/02_EDA-Zhuang/physeq_NZcomp.rds"))
```

```{r import-for-pdf, echo = FALSE}
physeq <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/OutputPhangorn/physeq_labus.rds")
physeq.rel <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/Labus_2017/Rproject_DataAnalysis/physeq_relative.rds")
physeq.clr <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/Labus_2017/Rproject_DataAnalysis/physeq_clr.rds")
metadata_table <- readRDS("/Users/scarcy/Desktop/Praktikum/Data/Labus_2017/metadata_final.rds")

saveRDS(physeq, "/Users/scarcy/Desktop/Praktikum/Data/Labus_2017/Rproject_DataAnalysis/physeq.rds")
# Distance methods
dist_methods <- unlist(distanceMethodList)
dist_methods <- dist_methods[-which(dist_methods=="ANY")] # Remove the user-defined distance
dist_methods <- dist_methods[-c(1,15,18)] #remove unweighted unifrac, mountford, chao (not gonna work on pseudocounts)
dist_methods <- dist_methods[1:16] #keep only the 16 first methods (remove the beta diversity distances)

# For heatmaps: have group color
mat_col <- data.frame(group = metadata_table[,"host_disease"], subtype = metadata_table[,"host_subtype"])
rownames(mat_col) <- rownames(metadata_table)
```



##################
## DEMOGRAPHICS ##
##################

This dataset has several covariates (gender, age, bmi). We will check whether there is the same distribution of these covariates between healthy and IBS patients.

```{r demographics-stats}
# Number of individuals in each group
metadata <- data.frame(sample_data(physeq.labus))
metadata %>%
  count(host_disease)

# Age
metadata %>%
  group_by(host_disease) %>%
  summarize(mean=mean(host_age), sd=sd(host_age))
wilcox.test(metadata[metadata$host_disease == "IBS", ]$host_age,
            metadata[metadata$host_disease == "Healthy", ]$host_age)

# Gender
metadata %>%
  count(host_disease, host_sex)
chisq.test(data.frame("Female" = c(21,14),
                      "Male" = c(8,9)))

# BMI
metadata %>%
  group_by(host_disease) %>%
  summarize(mean=mean(na.omit(host_bmi)), sd=sd(na.omit(host_bmi)))
wilcox.test(metadata[metadata$host_disease == "IBS",]$host_bmi,
            metadata[metadata$host_disease == "Healthy", ]$host_bmi)
```


################
## ABUNDANCES ##
################

### 1. Absolute abundances

```{r plot-absolute-abundance, fig.width=10, fig.height = 5}
# Plot Phylum
plot_bar(physeq.labus, fill = "Phylum") + facet_wrap("host_disease", scales="free") +
  theme(axis.text.x = element_text(size = 8))+
  labs(x = "Samples", y = "Absolute abundance", title = "Labus dataset (2017)") +
  ylim(0,7000)

# Plot Class
plot_bar(physeq.labus, fill = "Class")+ facet_wrap("host_disease", scales="free") +
  theme(axis.text.x = element_text(size = 8))+
  labs(x = "Samples", y = "Absolute abundance", title = "Labus dataset (2017)")+
  ylim(0,7000)
```

**Sequencing depth** characteristics of the Labus dataset:  
- minimum of `r min(sample_sums(physeq.labus))` total count per sample  
- median: `r median(sample_sums(physeq.labus))` total count per sample  
- maximum of `r max(sample_sums(physeq.labus))` total count per sample  

### 2. Relative abundances

```{r plot-relative-abundances, fig.width=10, fig.height = 5}
# Agglomerate to phylum & class levels
phylum.table <- physeq.labus %>%
  tax_glom(taxrank = "Phylum") %>%                     # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>% # Transform to rel. abundance
  psmelt()                                             # Melt to long format

class.table <- physeq.labus %>%
  tax_glom(taxrank = "Class") %>%
  transform_sample_counts(function(x) {x/sum(x)} ) %>%
  psmelt()


# Plot relative abundances
ggplot(phylum.table, aes(x = reorder(Sample, Sample, function(x) mean(phylum.table[Sample == x & Phylum == 'Firmicutes', 'Abundance'])),
                         y = Abundance, fill = Phylum))+
  facet_wrap(~ host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 8, angle = -90))+
  labs(x = "Samples", y = "Relative abundance", title = "Labus dataset (2017)")

ggplot(class.table, aes(x = reorder(Sample, Sample, function(x) mean(class.table[Sample == x & Phylum == 'Firmicutes', 'Abundance'])),
                        y = Abundance, fill = Class))+
  facet_wrap(~ host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 8, angle = -90))+
  labs(x = "Samples", y = "Relative abundance", title = "Labus dataset (2017)")

```


### 3. Firmicutes/Bacteroidota ratio

```{r plot-log-ratio, fig.height = 5, fig.width=3}
# Extract abundance of only Bacteroidota and Firmicutes
bacter <- phylum.table %>%
  filter(Phylum == "Bacteroidota") %>%
  select(c('Sample', 'Abundance', 'host_disease', 'host_subtype', 'Phylum', 'host_age', 'host_sex', 'host_bmi')) %>%
  arrange(Sample)

firmi <- phylum.table %>%
  filter(Phylum == "Firmicutes") %>%
  select(c('Sample', 'Abundance', 'host_disease', 'host_subtype', 'Phylum', 'host_age', 'host_sex', 'host_bmi')) %>%
  arrange(Sample)

# Calculate log2 ratio Firmicutes/Bacteroidota
ratio.FB <- data.frame('Sample' = bacter$Sample,
                       'host_disease' = bacter$host_disease,
                       'host_subtype' = bacter$host_subtype,
                       'Bacteroidota' = bacter$Abundance,
                       'Firmicutes' = firmi$Abundance,
                       'host_age' = bacter$host_age,
                       'host_sex' = bacter$host_sex,
                       'host_bmi' = bacter$host_bmi)
ratio.FB$logRatioFB <- log2(ratio.FB$Firmicutes / ratio.FB$Bacteroidota)

# Plot log2 ratio Firmicutes/Bacteroidota
ggplot(ratio.FB, aes(x = host_disease, y = logRatioFB))+
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(width=0.1)+
  labs(x = "",  y = 'Log2(Firmicutes/Bacteroidota)', title = "Firmicutes:Bacteroidota ratio")+
  theme_cowplot()

# Per IBS subtype
ggplot(ratio.FB, aes(x = host_subtype, y = logRatioFB))+
  geom_boxplot(outlier.shape = NA)+
  geom_jitter(width=0.1)+
  labs(x = "",  y = 'Log2(Firmicutes/Bacteroidota)', title = "")+
  theme_cowplot()
```


```{r export-plots-1, eval = FALSE, echo = FALSE}
# Save path to plots
path.plots <- file.path(path, "data/analysis-individual/PLOTS/plots-Labus-EDA")

# Absolute abundance phylum
jpeg(file.path(path.plots, "absAbundance_phylum.jpg"), width = 4000, height = 2000)
plot_bar(physeq.labus, fill = "Phylum") + facet_wrap("host_disease", scales="free") +
  theme(axis.text.x = element_text(size = 30, angle = -90),
        axis.text.y = element_text(size = 30),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70),
        strip.text = element_text(size = 50))+
  labs(x = "Samples", y = "Total read count", title = "Labus dataset (2017)")+
  ylim(0, 7000)
dev.off()

# Relative abundances
jpeg(file.path(path.plots, "relAbundance_phylum.jpg"), width = 4000, height = 2000)
ggplot(phylum.table, aes(x = Sample, y = Abundance, fill = Phylum))+
  facet_wrap(~ host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 30, angle = -90),
        axis.text.y = element_text(size = 30),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70),
        strip.text = element_text(size = 50))+
  labs(x = "Samples", y = "Relative abundance", title = "Labus dataset (2017)")
dev.off()

jpeg(file.path(path.plots, "relAbundance_class.jpg"), width = 4000, height = 2000)
ggplot(class.table, aes(x = Sample, y = Abundance, fill = Class))+
  facet_wrap(~ host_disease, scales = "free") + # scales = "free" removes empty lines
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(size = 30, angle = -90),
        axis.text.y = element_text(size = 30),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70),
        strip.text = element_text(size = 50))+
  labs(x = "Samples", y = "Relative abundance", title = "Labus dataset (2017)")
dev.off()

# Firmicutes:Bacteroidota log ratio
jpeg(file.path(path.plots, "ratioFB.jpg"), width = 1500, height = 2000)
ggplot(ratio.FB, aes(x = host_disease, y = logRatioFB))+
  geom_boxplot(outlier.size = 0, lwd = 3)+
  geom_jitter(size = 10, width=0.1)+
  labs(x = "",  y = 'Log2(Firmicutes/Bacteroidota)', title = "Labus dataset (2017)")+
  theme_classic()+
  theme(axis.text = element_text(size = 50),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 50))
dev.off()
```



####################
## NORMALIZE DATA ##
####################

```{r normalize-data, eval = FALSE, echo = TRUE}
# REMOVE SAMPLES WITH LESS THAN 1000 COUNTS
physeq_best <-  prune_samples(sample_sums(physeq)>=1000, physeq)
physeq_best

#____________________________________________________________________
# PHYLOSEQ OBJECT WITH NON-ZERO COMPOSITIONS
physeq.NZcomp <- physeq_best
otu_table(physeq.NZcomp)[otu_table(physeq.NZcomp) == 0] <- 0.5

# check the 0 values have been replaced
otu_table(physeq_best)[1:5,1:5]
otu_table(physeq.NZcomp)[1:5,1:5]

# transform into compositions
physeq.NZcomp <- transform_sample_counts(physeq.NZcomp, function(x) x / sum(x) ) # divide each count by the total number of counts (per sample)
sum(otu_table(physeq.NZcomp) < 0) # see how many negative values are present in the matrix
sum(rowSums(otu_table(physeq.NZcomp)) == 1) # check if there is any row not summing to 1

# save the physeq.NZcomp object
saveRDS(physeq.NZcomp, "/Users/scarcy/Desktop/Praktikum/Data/Labus_2017/Rproject_DataAnalysis/physeq_NZcomp.rds")

#____________________________________________________________________
# PHYLOSEQ OBJECT WITH RELATIVE COUNT (BETWEEN 0 AND 1)
physeq.rel <- physeq_best
physeq.rel <- transform_sample_counts(physeq.rel, function(x) x / sum(x) ) # divide each count by the total number of counts (per sample)

# check the counts are all relative
otu_table(physeq)[1:5, 1:5]
otu_table(physeq.rel)[1:5, 1:5]

# sanity check
sum(otu_table(physeq.rel) < 0) # see how many negative values are present in the matrix
sum(rowSums(otu_table(physeq.rel)) == 1) # check if there is any row not summing to 1

# save the physeq.rel object
saveRDS(physeq.rel, "/Users/scarcy/Desktop/Praktikum/Data/Labus_2017/Rproject_DataAnalysis/physeq_relative.rds")

#____________________________________________________________________
# PHYLOSEQ OBJECT WITH COMMON-SCALE NORMALIZATION
physeq.CSN <- physeq_best
physeq.CSN <- transform_sample_counts(physeq.CSN, function(x) (x*min(sample_sums(physeq.CSN))) / sum(x) ) # divide each count by the total number of counts (per sample)

# sanity check
sum(otu_table(physeq.CSN) < 0) # see how many negative values are present in the matrix
sum(rowSums(otu_table(physeq.CSN)) == 1024) # check if there is any row not summing to 1

# save the physeq.CSN object
saveRDS(physeq.CSN, "/Users/scarcy/Desktop/Praktikum/Data/Labus_2017/Rproject_DataAnalysis/physeq_CSN.rds")


#____________________________________________________________________
# PHYLOSEQ OBJECT WITH CENTERED LOG RATIO COUNT
physeq.clr <- physeq_best
physeq.clr <- transform(physeq.clr, "clr") # the function adds pseudocounts itself

# Compare the otu tables in the original phyloseq object and the new one after CLR transformation
otu_table(physeq_best)[1:5, 1:5] # should contain absolute counts
otu_table(physeq.clr)[1:5, 1:5] # should all be relative

# save the physeq.rel object
saveRDS(physeq.clr, "/Users/scarcy/Desktop/Praktikum/Data/Labus_2017/Rproject_DataAnalysis/physeq_clr.rds")
```



#######################
## COMPUTE DISTANCES ##
#######################

```{r dist-thesis, fig.width = 15, fig.height = 4}
#____________________________________________________________________________________
# Measure distances
Distances <- function(){
  set.seed(123) # for unifrac, need to set a seed
  glom.UniF <-  UniFrac(physeq.rel, weighted=TRUE, normalized=TRUE) # weighted unifrac
  glom.ait <- distance(physeq.clr, method = 'euclidean') # aitchison
  glom.bray <- distance(physeq.CSN, method = "bray") # bray-curtis
  glom.can <- distance(subset_samples(physeq.NZcomp, sample_names(physeq.NZcomp) != 'SRR5245605'),
                       method = "canberra") # canberra
  dist.list <- list("UniF" = glom.UniF, "Ait" = glom.ait, "Canb" = glom.can, "Bray" = glom.bray)
  
  return(dist.list)
}
#subset_samples(physeq.clr, sample_names(physeq.clr) != 'SRR5245605')

DistancesPlot <- function(d){
  plist <- NULL
  plist <- vector("list", 4)
  names(plist) <- c("Weighted Unifrac", "Aitchison", "Bray-Curtis", "Canberra")
  
  print("Unifrac")
  # Weighted UniFrac
  set.seed(123)
  iMDS.UniF <- ordinate(physeq.rel, "MDS", distance=d$UniF)
  plist[[1]] <- plot_ordination(physeq.rel, iMDS.UniF, color="host_disease")
  
  print("Aitchison")
  # Aitchison
  set.seed(123)
  iMDS.Ait <- ordinate(physeq.clr, "MDS", distance=d$Ait)
  plist[[2]] <- plot_ordination(physeq.clr, iMDS.Ait, color="host_disease")
  
  print("Bray")
  # Bray-Curtis
  set.seed(123)
  iMDS.Bray <- ordinate(physeq.CSN, "MDS", distance=d$Bray)
  plist[[3]] <- plot_ordination(physeq.CSN, iMDS.Bray, color="host_disease")
  
  print("Canberra")
  # Canberra
  set.seed(123)
  iMDS.Can <- ordinate(physeq.NZcomp, "MDS", distance=d$Can)
  plist[[4]] <- plot_ordination(physeq.NZcomp, iMDS.Can, color="host_disease")
  
  # Creating a dataframe to plot everything
  plot.df = ldply(plist, function(x) x$data)
  names(plot.df)[1] <- "distance"
  
  return(plot.df)
}

dist.thesis <- Distances()
plot.df <- DistancesPlot(dist.thesis)
plot.df$author <- "Labus"

p1 <- ggplot(plot.df[plot.df$distance == "Weighted Unifrac",], aes(Axis.1, Axis.2, color=host_disease))+
  geom_point(size=12, alpha=0.4)  + scale_color_manual(values = c('blue', 'red'))+
  facet_grid(author~distance, scales='free')+
  theme_bw()+
  theme(axis.title = element_blank(),
        #axis.title.x = element_text(size = 40),
        axis.text = element_text(size = 30),
        legend.text = element_text(size = 40),
        legend.title = element_text(size = 40),
        strip.background = element_rect(fill = "white", size = 3),
        panel.border = element_rect(size = 2),
        #strip.text.x = element_text(size = 50, margin = margin(1,0,1,0, "cm")),
        strip.text.x = element_blank(),
        strip.text.y = element_text(size = 70, margin = margin(0,1,0,1, "cm")),
        #axis.line = element_line(size = 1, arrow = arrow(length = unit(0.1, "inches")))
        )+
  labs(color="Disease")

p2 <- ggplot(plot.df[plot.df$distance != "Weighted Unifrac",], aes(Axis.1, Axis.2, color=host_disease))+
  geom_point(size=12, alpha=0.4)  + scale_color_manual(values = c('blue', 'red'))+
  facet_wrap(~distance, scales='free', nrow = 1)+
  guides(color = FALSE)+
  theme_bw()+
  theme(axis.text = element_text(size = 30),
        axis.title.y = element_text(size = 40),
        axis.title.x = element_blank(),
        strip.background = element_rect(fill = "white", size = 1),
        panel.border = element_rect(size = 2),
        #strip.text = element_text(size = 50, margin = margin(1,0,1,0, "cm")),
        strip.text.x = element_blank(),
        #axis.line = element_line(size = 1, arrow = arrow(length = unit(0.1, "inches")))
        )

library(ggplot2)
library(ggpubr)

jpeg("/Users/scarcy/Desktop/Praktikum/Thesis/dist_Labus_02.jpeg", height = 800, width = 4000, res = 100)
ggarrange(p2, p1, ncol = 2, widths = c(2,1))
dev.off()
```

```{r}
# Function that will allow to color the leaves of the tree according to sample type (biopsy vs stool)
color_leaf<<-function(n){
    if(is.leaf(n)){
        # take the current attributes
        a=attributes(n)
        
        # deduce the line in the original data, to get the corresponding sampleType status
        line=match(attributes(n)$label, metadata_table$Run)
        disease=metadata_table[line, "host_disease"];
            if(disease=="Healthy"){col_disease="blue"};if(disease=="IBS"){col_disease="red"}
        
        #Modification of leaf attribute
        attr(n,"nodePar") <- c(a$nodePar, list(cex=1.5, lab.cex=0.6, pch=20, col=col_disease, lab.font=1, lab.cex=1))
        }
    return(n)
}

```

```{r, fig.height = 5, fig.width = 3}
hc.ward <- hclust(no_glom.dist$Canb, method = "ward.D") #clustering 
dend.ward <- as.dendrogram(hc.ward) # Make into dendograms

dend.ward <- dendrapply(dend.ward, color_leaf) # Apply the color attributes to the dendograms

jpeg("/Users/scarcy/Desktop/Praktikum/Thesis/hclust_labus.jpg", height = 2000, width = 1000, res = 200)
plot(dend.ward , main="Hierarchical clustering on Canberra distance (ward linkage)", horiz = TRUE, leaflab = "none",
     edgePar = list(lwd = 2))
axis(side = 1, lwd = 4)
legend(4, 50, 
     legend = c("Healthy" , "IBS"), 
     col = c("blue", "red"), 
     pch = c(20,20), bty = "n",  pt.cex = 1.5, cex = 1, 
     text.col = "black", horiz = FALSE, inset = c(0, 0.1))
dev.off()
```


```{r dist-functions, eval = TRUE, echo = TRUE}
#____________________________________________________________________________________
# Measure distances
Distances <- function(physeq_obj){
  set.seed(123) # for unifrac, need to set a seed
  glom.UniF <-  UniFrac(physeq_obj, weighted=TRUE, normalized=TRUE) # weighted unifrac
  glom.ait <- distance(subset_samples(physeq.clr, sample_names(physeq.clr) != 'SRR5245605'), method = 'euclidean') # aitchison
  glom.bray <- distance(physeq_obj, method = "bray") # bray-curtis
  glom.jac <- distance(physeq_obj, method = "jaccard") # jaccard
  glom.can <- distance(physeq_obj, method = "canberra") # canberra
  dist.list <- list("UniF" = glom.UniF, "Ait" = glom.ait, "Canb" = glom.can, "Bray" = glom.bray, "Jac" = glom.jac)
  
  return(dist.list)
}

#____________________________________________________________________________________
# Plot 2D ordination
MDS_2D <- function(physeq_obj, ait_dist){
  
  plist <- NULL
  plist <- vector("list", length(dist_methods)+1) # save each plot to a list
  names(plist) <- dist_methods # save the name of each method
  names(plist)[17] <- "aitchison" # save the name of aitchison
  
  # Loop through all distance methods
  for(i in dist_methods){
    # Calculate distance matrix
    #print(i)
    set.seed(123) # in case the distance method needs a rooted tree (weighted unifrac)
    iDist <- phyloseq::distance(physeq_obj, method=i)
    # Calculate ordination
    set.seed(123)
    iMDS  <- ordinate(physeq_obj, "MDS", distance=iDist)
    ## Make plot
    # Don't carry over previous plot (if error, p will be blank)
    p <- NULL
    # Create plot, store as temporary variable, p
    p <- plot_ordination(physeq_obj, iMDS, color="host_disease")
    # Add title to each plot
    p <- p + ggtitle(paste("MDS using distance method ", i, sep=""))
    # Save the graphic to the plot list
    plist[[i]] = p
  }
  
  # Add aitchison
  iMDS  <- ordinate(physeq_obj, "MDS", distance=ait_dist)
  p <- NULL
  p <- plot_ordination(physeq_obj, iMDS, color="host_disease")
  p <- p + ggtitle("MDS using distance method Aitchison")
  plist[[17]] = p
  
  # Creating a dataframe to plot everything
  plot.df = ldply(plist, function(x) x$data)
  names(plot.df)[1] <- "distance"
  
  # Plot
  p.alldist <- ggplot(plot.df, aes(Axis.1, Axis.2, color=host_disease))+
    geom_point(size=5, alpha=0.5)  + scale_color_manual(values = c('blue', 'red'))+
    facet_wrap(~distance, scales='free')+
    theme(strip.text = element_text(size = 40),
          legend.text = element_text(size = 20),
          axis.text.x = element_text(size = 20),
          axis.text.y = element_text(size = 20))+
    geom_text_repel(data=plot.df %>% filter(Run == 'SRR5245605'), # Filter data first
                    aes(label='SRR5245605'), size = 5, nudge_y = -0.1, nudge_x = -0.1)

  return(p.alldist)
}

#____________________________________________________________________________________
# Plot 3D ordination
MDS_3D <- function(d, name_dist){
  
  # Reset parameters
  mds.3D <- NULL
  xyz <- NULL
  fig.3D <- NULL
  
  # Reduce distance matrix to 3 dimensions
  set.seed(123) # to get the same dimensionality reduction at each run
  mds.3D <- metaMDS(d, method="MDS", k=3, trace = 0)
  xyz <- scores(mds.3D, display="sites") # pull out the x y z coordinates
  
  if(name_dist == 'Aitchison'){
    fig.3D <- plot_ly(x=xyz[,1], y=xyz[,2], z=xyz[,3], type='scatter3d', mode='markers',
                      color=sample_data(subset_samples(physeq.rel, sample_names(physeq.rel) != 'SRR5245605'))$host_disease,
                      colors = c('blue', 'red')) %>%
    layout(title = paste('MDS in 3D with', name_dist, 'distance', sep = ' '))
  }
  
  else{
    fig.3D <- plot_ly(x=xyz[,1], y=xyz[,2], z=xyz[,3], type='scatter3d',
                       mode='markers', color=sample_data(physeq.rel)$host_disease, colors = c('blue', 'red')) %>%
      add_trace(x=xyz['SRR5245605',1], y=xyz['SRR5245605',2], z=xyz['SRR5245605',3],
            type='scatter3d', mode='text', text = 'SRR5245605', textfont = list(color = 'blue')) %>%
      layout(title = paste('MDS in 3D with', name_dist, 'distance', sep = ' '))
  }
  
  return(fig.3D)
}

```


```{r no-glom1, fig.height = 20, fig.width = 30}
# Get the distances
no_glom.dist <- Distances(physeq_obj = physeq.rel)

# Get the 2D ordination plots
no_glom.2D.alldist <- MDS_2D(physeq.rel, no_glom.dist$Ait)
no_glom.2D.alldist + theme(title = element_text(size = 30))

```


```{r save-distances, eval = FALSE, echo = FALSE}
getwd()

write.csv(as.matrix(no_glom.dist$UniF), file = "dist_wunifrac.csv", row.names = TRUE)
write.csv(as.matrix(no_glom.dist$Bray), file = "dist_brayCurtis.csv", row.names = TRUE)
write.csv(as.matrix(no_glom.dist$Ait), file = "dist_aitchison.csv", row.names = TRUE)
write.csv(as.matrix(no_glom.dist$Canb), file = "dist_canberra.csv", row.names = TRUE)
```



```{r no-glom2, fig.height = 6, fig.width = 10, message = FALSE, warning = FALSE}
# Get 3D MDS plots
no_glom.3D.UniF <- MDS_3D(no_glom.dist$UniF, 'weighted Unifrac')
no_glom.3D.Ait <- MDS_3D(no_glom.dist$Ait, 'Aitchison')
no_glom.3D.Jac <- MDS_3D(no_glom.dist$Jac, 'Jaccard')
no_glom.3D.Can <- MDS_3D(no_glom.dist$Canb, 'Canberra')

no_glom.3D.UniF
no_glom.3D.Ait
no_glom.3D.Jac
no_glom.3D.Can
```



#############################
## HIERARCHICAL CLUSTERING ##
#############################

```{r hierarchical-clust, fig.width = 10, fig.width = 7}
Heatmaps <- function(dist_list, fontsize){
  # Weighted Unifrac
  heatmp.UniF <- pheatmap(as.matrix(dist_list$UniF), 
                          clustering_distance_rows = dist_list$UniF,
                          clustering_distance_cols = dist_list$UniF,
                          fontsize = fontsize,
                          fontsize_col = fontsize-5,
                          fontsize_row = fontsize-5,
                          annotation_col = mat_col,
                          annotation_row = mat_col,
                          annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                   subtype = c('HC' = 'blue',
                                                               'IBS-D' = '#CC0033', 'IBS-C' = '#990000', 'IBS-M' = '#FF0099',
                                                               'IBS-unspecified' = '#FF6666', 'IBS-A' = '#FF0099')),
                          cluster_rows = T,
                          cluster_cols = T,
                          clustering_method = 'complete', #hierarchical method
                          main = 'Weighted unifrac distance')

  # Aitchison
  heatmp.Ait <- pheatmap(as.matrix(dist_list$Ait), 
                         clustering_distance_rows = dist_list$Ait,
                         clustering_distance_cols = dist_list$Ait,
                         fontsize = fontsize,
                         fontsize_col = fontsize-5,
                         fontsize_row = fontsize-5,
                         # border_color = NA,
                         cluster_rows = T,
                         cluster_cols = T,
                         clustering_method = "complete", #hierarchical method
                         annotation_col = mat_col,
                         annotation_row = mat_col,
                         annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                   subtype = c('HC' = 'blue',
                                                               'IBS-D' = '#CC0033', 'IBS-C' = '#990000', 'IBS-M' = '#FF0099',
                                                               'IBS-unspecified' = '#FF6666', 'IBS-A' = '#FF0099')),
                         main = "Aitchison distance")
  
  # Bray-Curtis
  heatmp.Bray <- pheatmap(as.matrix(dist_list$Bray), 
                         clustering_distance_rows = dist_list$Bray,
                         clustering_distance_cols = dist_list$Bray,
                         fontsize = fontsize,
                         fontsize_col = fontsize-5,
                         fontsize_row = fontsize-5,
                         # border_color = NA,
                         cluster_rows = T,
                         cluster_cols = T,
                         clustering_method = "complete", #hierarchical method
                         annotation_col = mat_col,
                         annotation_row = mat_col,
                         annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                   subtype = c('HC' = 'blue',
                                                               'IBS-D' = '#CC0033', 'IBS-C' = '#990000', 'IBS-M' = '#FF0099',
                                                               'IBS-unspecified' = '#FF6666', 'IBS-A' = '#FF0099')),
                         main = "Bray-Curtis distance")
  
  # Canberra
  heatmp.Canb <- pheatmap(as.matrix(dist_list$Canb), 
                         clustering_distance_rows = dist_list$Canb,
                         clustering_distance_cols = dist_list$Canb,
                         fontsize = fontsize,
                         fontsize_col = fontsize-5,
                         fontsize_row = fontsize-5,
                         # border_color = NA,
                         cluster_rows = T,
                         cluster_cols = T,
                         clustering_method = "complete", #hierarchical method
                         annotation_col = mat_col,
                         annotation_row = mat_col,
                         annotation_colors = list(group = c('Healthy' = 'blue', 'IBS' = 'red'),
                                                   subtype = c('HC' = 'blue',
                                                               'IBS-D' = '#CC0033', 'IBS-C' = '#990000', 'IBS-M' = '#FF0099',
                                                               'IBS-unspecified' = '#FF6666', 'IBS-A' = '#FF0099')),
                         main = "Canberra distance")
  
  return(list("UniF" = heatmp.UniF, "Ait" = heatmp.Ait, "Bray" = heatmp.Bray, "Canb" = heatmp.Canb))
}


# Get the heatmaps
no_glom.dist$Canb <- distance(subset_samples(physeq.rel, sample_names(physeq.rel) != 'SRR5245605'), method = "canberra") # remove the outlier in the canberra distance
no_glom.heatmaps <- Heatmaps(dist_list = no_glom.dist, fontsize = 8)
```



```{r export-plots-2, eval = FALSE, echo = FALSE}
# MDS 2D all distances
jpeg("./Plots/AllDistances_MDS.jpg", width = 3000, height = 2000, res = 90)
no_glom.2D.alldist +
  theme(axis.text.x = element_text(size = 20),
        legend.text = element_text(size = 50),
        axis.title = element_text(size = 50),
        title = element_text(size = 70),
        strip.text = element_text(size = 50))+
  labs(title = "Labus dataset (2017)")
dev.off()

# Heatmaps
jpeg("./Plots/Heatmap_Aitchison.jpg", width = 1400, height = 1000, res = 200)
no_glom.heatmaps$Ait
dev.off()

jpeg("./Plots/Heatmap_UniF.jpg", width = 1400, height = 1000, res = 200)
no_glom.heatmaps$UniF
dev.off()

jpeg("./Plots/Heatmap_BrayCurtis.jpg", width = 1400, height = 1000, res = 200)
no_glom.heatmaps$Bray
dev.off()

jpeg("./Plots/Heatmap_Canberra.jpg", width = 1400, height = 1000, res = 200)
no_glom.heatmaps$Can
dev.off()

```


################################
## REPRODUCE PLOTS FROM PAPER ##
################################

```{r reproduce-plots, eval = TRUE, echo = TRUE, fig.height = 6, fig.width = 10}
#___________________________________________________________________
# FIGURE 1A
set.seed(785)
UniF.test = UniFrac(physeq, weighted=FALSE, normalized=T) # unweighted Unifrac dist

UniF.PCoA <- ordinate(physeq, "PCoA", distance = UniF.test) # PCoA on unweighted unifrac dist
plot_ordination(physeq, UniF.PCoA, color="host_disease")

# Plotting the first 3 dimensions of the PCoA
xyz <- as.matrix(UniF.PCoA$vectors[,1:3])
test.plot <- plot_ly(x=xyz[,1], y=xyz[,2], z=xyz[,3], type="scatter3d",
                       mode="markers", color=sample_data(physeq)$host_disease, colors = c("blue", "red"))%>%
  layout(title = "PCoA in 3D with unweighted Unifrac distance")
test.plot


#___________________________________________________________________
# FIGURE 1B

# Hierarchical clustering using average linkage
hc.average <- hclust(UniF.test, method = "average")
dend.average <- as.dendrogram(hc.average)
color_leaf<<-function(n){
    if(is.leaf(n)){
        # take the current attributes
        a=attributes(n)
        
        # deduce the line in the original data, to get the corresponding disease status
        line=match(attributes(n)$label, rownames(metadata_table))
        disease=metadata_table[line, "host_disease"];
            if(disease=="Healthy"){col_disease="blue"};if(disease=="IBS"){col_disease="red"}
        
        #Modification of leaf attribute
        attr(n,"nodePar") <- c(a$nodePar, list(cex=1.5, lab.cex=0.6, pch=20, col=col_disease, lab.font=1, lab.cex=0.1))
        }
    return(n)
}
dend.average <- dendrapply(dend.average, color_leaf)

#Plot
par(mar=c(3,2,3,7), xpd=T, cex = 0.8)
plot(dend.average , main="Hierarchical clustering: average linkage", horiz = T, cex = 0.1)
legend('topleft',
     legend = c("Healthy" , "IBS"), 
     col = c("blue", "red"), 
     pch = c(20,20), bty = "n",  pt.cex = 1.5, cex = 1, 
     text.col = "black", horiz = FALSE, inset = c(0, 0.1))
```


